<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 13 Examples of algorithms | Preceptor’s Primer for Bayesian Data Science</title>
  <meta name="description" content="Chapter 13 Examples of algorithms | Preceptor’s Primer for Bayesian Data Science" />
  <meta name="generator" content="bookdown 0.14 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 13 Examples of algorithms | Preceptor’s Primer for Bayesian Data Science" />
  <meta property="og:type" content="book" />
  
  
  
  <meta name="github-repo" content="davidkane9/PPBDS" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 13 Examples of algorithms | Preceptor’s Primer for Bayesian Data Science" />
  
  
  

<meta name="author" content="David Kane" />


<meta name="date" content="2019-12-30" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="12-cross-validation.html"/>
<link rel="next" href="14-thinking-with-data.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<script src="libs/htmlwidgets-1.5.1/htmlwidgets.js"></script>
<link href="libs/dygraphs-1.1.1/dygraph.css" rel="stylesheet" />
<script src="libs/dygraphs-1.1.1/dygraph-combined.js"></script>
<script src="libs/dygraphs-1.1.1/shapes.js"></script>
<script src="libs/moment-2.8.4/moment.js"></script>
<script src="libs/moment-timezone-0.2.5/moment-timezone-with-data.js"></script>
<script src="libs/moment-fquarter-1.0.0/moment-fquarter.min.js"></script>
<script src="libs/dygraphs-binding-1.1.1.6/dygraphs.js"></script>


<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Forward</a></li>
<li class="chapter" data-level="" data-path="acknowledgements.html"><a href="acknowledgements.html"><i class="fa fa-check"></i>Acknowledgements</a></li>
<li class="chapter" data-level="1" data-path="1-getting-started.html"><a href="1-getting-started.html"><i class="fa fa-check"></i><b>1</b> Getting Started with Data in R</a><ul>
<li class="chapter" data-level="1.1" data-path="1-getting-started.html"><a href="1-getting-started.html#r-rstudio"><i class="fa fa-check"></i><b>1.1</b> What are R and RStudio?</a><ul>
<li class="chapter" data-level="1.1.1" data-path="1-getting-started.html"><a href="1-getting-started.html#installing"><i class="fa fa-check"></i><b>1.1.1</b> Installing R and RStudio</a></li>
<li class="chapter" data-level="1.1.2" data-path="1-getting-started.html"><a href="1-getting-started.html#using-r-via-rstudio"><i class="fa fa-check"></i><b>1.1.2</b> Using R via RStudio</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="1-getting-started.html"><a href="1-getting-started.html#code"><i class="fa fa-check"></i><b>1.2</b> How do I code in R?</a><ul>
<li class="chapter" data-level="1.2.1" data-path="1-getting-started.html"><a href="1-getting-started.html#programming-concepts"><i class="fa fa-check"></i><b>1.2.1</b> Basic programming concepts and terminology</a></li>
<li class="chapter" data-level="1.2.2" data-path="1-getting-started.html"><a href="1-getting-started.html#messages"><i class="fa fa-check"></i><b>1.2.2</b> Errors, warnings, and messages</a></li>
<li class="chapter" data-level="1.2.3" data-path="1-getting-started.html"><a href="1-getting-started.html#tips-code"><i class="fa fa-check"></i><b>1.2.3</b> Tips on learning to code</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="1-getting-started.html"><a href="1-getting-started.html#packages"><i class="fa fa-check"></i><b>1.3</b> What are R packages?</a><ul>
<li class="chapter" data-level="1.3.1" data-path="1-getting-started.html"><a href="1-getting-started.html#package-installation"><i class="fa fa-check"></i><b>1.3.1</b> Package installation</a></li>
<li class="chapter" data-level="1.3.2" data-path="1-getting-started.html"><a href="1-getting-started.html#package-loading"><i class="fa fa-check"></i><b>1.3.2</b> Package loading</a></li>
<li class="chapter" data-level="1.3.3" data-path="1-getting-started.html"><a href="1-getting-started.html#package-use"><i class="fa fa-check"></i><b>1.3.3</b> Package use</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="1-getting-started.html"><a href="1-getting-started.html#nycflights13"><i class="fa fa-check"></i><b>1.4</b> Explore your first datasets</a><ul>
<li class="chapter" data-level="1.4.1" data-path="1-getting-started.html"><a href="1-getting-started.html#nycflights13-package"><i class="fa fa-check"></i><b>1.4.1</b> <code>nycflights13</code> package</a></li>
<li class="chapter" data-level="1.4.2" data-path="1-getting-started.html"><a href="1-getting-started.html#flights-data-frame"><i class="fa fa-check"></i><b>1.4.2</b> <code>flights</code> data frame</a></li>
<li class="chapter" data-level="1.4.3" data-path="1-getting-started.html"><a href="1-getting-started.html#exploredataframes"><i class="fa fa-check"></i><b>1.4.3</b> Exploring data frames</a></li>
<li class="chapter" data-level="1.4.4" data-path="1-getting-started.html"><a href="1-getting-started.html#identification-vs-measurement-variables"><i class="fa fa-check"></i><b>1.4.4</b> Identification and measurement variables</a></li>
<li class="chapter" data-level="1.4.5" data-path="1-getting-started.html"><a href="1-getting-started.html#help-files"><i class="fa fa-check"></i><b>1.4.5</b> Help files</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="1-getting-started.html"><a href="1-getting-started.html#conclusion"><i class="fa fa-check"></i><b>1.5</b> Conclusion</a><ul>
<li class="chapter" data-level="1.5.1" data-path="1-getting-started.html"><a href="1-getting-started.html#additional-resources"><i class="fa fa-check"></i><b>1.5.1</b> Additional resources</a></li>
<li class="chapter" data-level="1.5.2" data-path="1-getting-started.html"><a href="1-getting-started.html#whats-to-come"><i class="fa fa-check"></i><b>1.5.2</b> What’s to come?</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="2" data-path="2-viz.html"><a href="2-viz.html"><i class="fa fa-check"></i><b>2</b> Data Visualization</a><ul>
<li class="chapter" data-level="" data-path="2-viz.html"><a href="2-viz.html#needed-packages"><i class="fa fa-check"></i>Needed packages</a></li>
<li class="chapter" data-level="2.1" data-path="2-viz.html"><a href="2-viz.html#grammarofgraphics"><i class="fa fa-check"></i><b>2.1</b> The grammar of graphics</a><ul>
<li class="chapter" data-level="2.1.1" data-path="2-viz.html"><a href="2-viz.html#components-of-the-grammar"><i class="fa fa-check"></i><b>2.1.1</b> Components of the grammar</a></li>
<li class="chapter" data-level="2.1.2" data-path="2-viz.html"><a href="2-viz.html#gapminder"><i class="fa fa-check"></i><b>2.1.2</b> Gapminder data</a></li>
<li class="chapter" data-level="2.1.3" data-path="2-viz.html"><a href="2-viz.html#other-components"><i class="fa fa-check"></i><b>2.1.3</b> Other components</a></li>
<li class="chapter" data-level="2.1.4" data-path="2-viz.html"><a href="2-viz.html#ggplot2-package"><i class="fa fa-check"></i><b>2.1.4</b> ggplot2 package</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="2-viz.html"><a href="2-viz.html#FiveNG"><i class="fa fa-check"></i><b>2.2</b> Five named graphs - the 5NG</a></li>
<li class="chapter" data-level="2.3" data-path="2-viz.html"><a href="2-viz.html#scatterplots"><i class="fa fa-check"></i><b>2.3</b> 5NG#1: Scatterplots</a><ul>
<li class="chapter" data-level="2.3.1" data-path="2-viz.html"><a href="2-viz.html#geompoint"><i class="fa fa-check"></i><b>2.3.1</b> Scatterplots via <code>geom_point</code></a></li>
<li class="chapter" data-level="2.3.2" data-path="2-viz.html"><a href="2-viz.html#overplotting"><i class="fa fa-check"></i><b>2.3.2</b> Overplotting</a></li>
<li class="chapter" data-level="2.3.3" data-path="2-viz.html"><a href="2-viz.html#summary"><i class="fa fa-check"></i><b>2.3.3</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="2-viz.html"><a href="2-viz.html#linegraphs"><i class="fa fa-check"></i><b>2.4</b> 5NG#2: Linegraphs</a><ul>
<li class="chapter" data-level="2.4.1" data-path="2-viz.html"><a href="2-viz.html#geomline"><i class="fa fa-check"></i><b>2.4.1</b> Linegraphs via <code>geom_line</code></a></li>
<li class="chapter" data-level="2.4.2" data-path="2-viz.html"><a href="2-viz.html#summary-1"><i class="fa fa-check"></i><b>2.4.2</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="2-viz.html"><a href="2-viz.html#histograms"><i class="fa fa-check"></i><b>2.5</b> 5NG#3: Histograms</a><ul>
<li class="chapter" data-level="2.5.1" data-path="2-viz.html"><a href="2-viz.html#geomhistogram"><i class="fa fa-check"></i><b>2.5.1</b> Histograms via <code>geom_histogram</code></a></li>
<li class="chapter" data-level="2.5.2" data-path="2-viz.html"><a href="2-viz.html#adjustbins"><i class="fa fa-check"></i><b>2.5.2</b> Adjusting the bins</a></li>
<li class="chapter" data-level="2.5.3" data-path="2-viz.html"><a href="2-viz.html#summary-2"><i class="fa fa-check"></i><b>2.5.3</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="2-viz.html"><a href="2-viz.html#facets"><i class="fa fa-check"></i><b>2.6</b> Facets</a></li>
<li class="chapter" data-level="2.7" data-path="2-viz.html"><a href="2-viz.html#boxplots"><i class="fa fa-check"></i><b>2.7</b> 5NG#4: Boxplots</a><ul>
<li class="chapter" data-level="2.7.1" data-path="2-viz.html"><a href="2-viz.html#geomboxplot"><i class="fa fa-check"></i><b>2.7.1</b> Boxplots via <code>geom_boxplot</code></a></li>
<li class="chapter" data-level="2.7.2" data-path="2-viz.html"><a href="2-viz.html#summary-3"><i class="fa fa-check"></i><b>2.7.2</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="2-viz.html"><a href="2-viz.html#geombar"><i class="fa fa-check"></i><b>2.8</b> 5NG#5: Barplots</a><ul>
<li class="chapter" data-level="2.8.1" data-path="2-viz.html"><a href="2-viz.html#barplots-via-geom_bar-or-geom_col"><i class="fa fa-check"></i><b>2.8.1</b> Barplots via <code>geom_bar</code> or <code>geom_col</code></a></li>
<li class="chapter" data-level="2.8.2" data-path="2-viz.html"><a href="2-viz.html#must-avoid-pie-charts"><i class="fa fa-check"></i><b>2.8.2</b> Must avoid pie charts!</a></li>
<li class="chapter" data-level="2.8.3" data-path="2-viz.html"><a href="2-viz.html#two-categ-barplot"><i class="fa fa-check"></i><b>2.8.3</b> Two categorical variables</a></li>
<li class="chapter" data-level="2.8.4" data-path="2-viz.html"><a href="2-viz.html#summary-4"><i class="fa fa-check"></i><b>2.8.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="2.9" data-path="2-viz.html"><a href="2-viz.html#conclusion-1"><i class="fa fa-check"></i><b>2.9</b> Conclusion</a><ul>
<li class="chapter" data-level="2.9.1" data-path="2-viz.html"><a href="2-viz.html#summary-table"><i class="fa fa-check"></i><b>2.9.1</b> Summary table</a></li>
<li class="chapter" data-level="2.9.2" data-path="2-viz.html"><a href="2-viz.html#function-argument-specification"><i class="fa fa-check"></i><b>2.9.2</b> Function argument specification</a></li>
<li class="chapter" data-level="2.9.3" data-path="2-viz.html"><a href="2-viz.html#additional-resources-1"><i class="fa fa-check"></i><b>2.9.3</b> Additional resources</a></li>
<li class="chapter" data-level="2.9.4" data-path="2-viz.html"><a href="2-viz.html#whats-to-come-3"><i class="fa fa-check"></i><b>2.9.4</b> What’s to come</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html"><i class="fa fa-check"></i><b>3</b> Productivity Tools</a><ul>
<li class="chapter" data-level="3.1" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#accessing-the-terminal-and-installing-git"><i class="fa fa-check"></i><b>3.1</b> Accessing the terminal and installing Git</a></li>
<li class="chapter" data-level="3.2" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#terminal-on-mac"><i class="fa fa-check"></i><b>3.2</b> Accessing the terminal on a Mac</a></li>
<li class="chapter" data-level="3.3" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#installing-git-on-the-mac"><i class="fa fa-check"></i><b>3.3</b> Installing Git on the Mac</a></li>
<li class="chapter" data-level="3.4" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#installing-git-and-git-bash-on-windows"><i class="fa fa-check"></i><b>3.4</b> Installing Git and Git Bash on Windows</a></li>
<li class="chapter" data-level="3.5" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#terminal-on-windows"><i class="fa fa-check"></i><b>3.5</b> Accessing the terminal on Windows</a></li>
<li class="chapter" data-level="3.6" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#unix"><i class="fa fa-check"></i><b>3.6</b> Organizing with Unix</a></li>
<li class="chapter" data-level="3.7" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#naming-convention"><i class="fa fa-check"></i><b>3.7</b> Naming convention</a></li>
<li class="chapter" data-level="3.8" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#the-terminal"><i class="fa fa-check"></i><b>3.8</b> The terminal</a></li>
<li class="chapter" data-level="3.9" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#filesystem"><i class="fa fa-check"></i><b>3.9</b> The filesystem</a><ul>
<li class="chapter" data-level="3.9.1" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#directories-and-subdirectories"><i class="fa fa-check"></i><b>3.9.1</b> Directories and subdirectories</a></li>
<li class="chapter" data-level="3.9.2" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#the-home-directory"><i class="fa fa-check"></i><b>3.9.2</b> The home directory</a></li>
<li class="chapter" data-level="3.9.3" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#working-directory"><i class="fa fa-check"></i><b>3.9.3</b> Working directory</a></li>
<li class="chapter" data-level="3.9.4" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#paths"><i class="fa fa-check"></i><b>3.9.4</b> Paths</a></li>
</ul></li>
<li class="chapter" data-level="3.10" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#unix-commands"><i class="fa fa-check"></i><b>3.10</b> Unix commands</a><ul>
<li class="chapter" data-level="3.10.1" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#ls-listing-directory-content"><i class="fa fa-check"></i><b>3.10.1</b> <code>ls</code>: Listing directory content</a></li>
<li class="chapter" data-level="3.10.2" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#mkdir-and-rmdir-make-and-remove-a-directory"><i class="fa fa-check"></i><b>3.10.2</b> <code>mkdir</code> and <code>rmdir</code>: make and remove a directory</a></li>
<li class="chapter" data-level="3.10.3" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#cd-navigating-the-filesystem-by-changing-directories"><i class="fa fa-check"></i><b>3.10.3</b> <code>cd</code>: navigating the filesystem by changing directories</a></li>
</ul></li>
<li class="chapter" data-level="3.11" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#some-examples"><i class="fa fa-check"></i><b>3.11</b> Some examples</a></li>
<li class="chapter" data-level="3.12" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#more-unix-commands"><i class="fa fa-check"></i><b>3.12</b> More Unix commands</a><ul>
<li class="chapter" data-level="3.12.1" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#mv-moving-files"><i class="fa fa-check"></i><b>3.12.1</b> <code>mv</code>: moving files</a></li>
<li class="chapter" data-level="3.12.2" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#cp-copying-files"><i class="fa fa-check"></i><b>3.12.2</b> <code>cp</code>: copying files</a></li>
<li class="chapter" data-level="3.12.3" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#rm-removing-files"><i class="fa fa-check"></i><b>3.12.3</b> <code>rm</code>: removing files</a></li>
<li class="chapter" data-level="3.12.4" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#less-looking-at-a-file"><i class="fa fa-check"></i><b>3.12.4</b> <code>less</code>: looking at a file</a></li>
</ul></li>
<li class="chapter" data-level="3.13" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#prep-project"><i class="fa fa-check"></i><b>3.13</b> Preparing for a data science project</a></li>
<li class="chapter" data-level="3.14" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#advanced-unix"><i class="fa fa-check"></i><b>3.14</b> Advanced Unix</a><ul>
<li class="chapter" data-level="3.14.1" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#arguments"><i class="fa fa-check"></i><b>3.14.1</b> Arguments</a></li>
<li class="chapter" data-level="3.14.2" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#getting-help"><i class="fa fa-check"></i><b>3.14.2</b> Getting help</a></li>
<li class="chapter" data-level="3.14.3" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#pipes"><i class="fa fa-check"></i><b>3.14.3</b> Pipes</a></li>
<li class="chapter" data-level="3.14.4" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#wild-cards"><i class="fa fa-check"></i><b>3.14.4</b> Wild cards</a></li>
<li class="chapter" data-level="3.14.5" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#environment-variables"><i class="fa fa-check"></i><b>3.14.5</b> Environment variables</a></li>
<li class="chapter" data-level="3.14.6" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#shells"><i class="fa fa-check"></i><b>3.14.6</b> Shells</a></li>
<li class="chapter" data-level="3.14.7" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#executables"><i class="fa fa-check"></i><b>3.14.7</b> Executables</a></li>
<li class="chapter" data-level="3.14.8" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#permissions-and-file-types"><i class="fa fa-check"></i><b>3.14.8</b> Permissions and file types</a></li>
<li class="chapter" data-level="3.14.9" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#commands-you-should-learn"><i class="fa fa-check"></i><b>3.14.9</b> Commands you should learn</a></li>
<li class="chapter" data-level="3.14.10" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#file-manipulation-in-r"><i class="fa fa-check"></i><b>3.14.10</b> File manipulation in R</a></li>
</ul></li>
<li class="chapter" data-level="3.15" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#git"><i class="fa fa-check"></i><b>3.15</b> Git and GitHub</a></li>
<li class="chapter" data-level="3.16" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#why-use-git-and-github"><i class="fa fa-check"></i><b>3.16</b> Why use Git and GitHub?</a></li>
<li class="chapter" data-level="3.17" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#github-accounts"><i class="fa fa-check"></i><b>3.17</b> GitHub accounts</a></li>
<li class="chapter" data-level="3.18" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#github-repos"><i class="fa fa-check"></i><b>3.18</b> GitHub repositories</a></li>
<li class="chapter" data-level="3.19" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#git-overview"><i class="fa fa-check"></i><b>3.19</b> Overview of Git</a><ul>
<li class="chapter" data-level="3.19.1" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#clone"><i class="fa fa-check"></i><b>3.19.1</b> Clone</a></li>
</ul></li>
<li class="chapter" data-level="3.20" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#init"><i class="fa fa-check"></i><b>3.20</b> Initializing a Git directory</a></li>
<li class="chapter" data-level="3.21" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#rstudio-git"><i class="fa fa-check"></i><b>3.21</b> Using Git and GitHub in RStudio</a></li>
<li class="chapter" data-level="3.22" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#reproducible-projects-with-rstudio-and-r-markdown"><i class="fa fa-check"></i><b>3.22</b> Reproducible projects with RStudio and R markdown</a></li>
<li class="chapter" data-level="3.23" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#rstudio-projects"><i class="fa fa-check"></i><b>3.23</b> RStudio projects</a></li>
<li class="chapter" data-level="3.24" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#r-markdown"><i class="fa fa-check"></i><b>3.24</b> R markdown</a><ul>
<li class="chapter" data-level="3.24.1" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#the-header"><i class="fa fa-check"></i><b>3.24.1</b> The header</a></li>
<li class="chapter" data-level="3.24.2" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#r-code-chunks"><i class="fa fa-check"></i><b>3.24.2</b> R code chunks</a></li>
<li class="chapter" data-level="3.24.3" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#global-options"><i class="fa fa-check"></i><b>3.24.3</b> Global options</a></li>
<li class="chapter" data-level="3.24.4" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#knitr"><i class="fa fa-check"></i><b>3.24.4</b> knitR</a></li>
<li class="chapter" data-level="3.24.5" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#more-on-r-markdown"><i class="fa fa-check"></i><b>3.24.5</b> More on R markdown</a></li>
</ul></li>
<li class="chapter" data-level="3.25" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#organizing"><i class="fa fa-check"></i><b>3.25</b> Organizing a data science project</a><ul>
<li class="chapter" data-level="3.25.1" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#create-directories-in-unix"><i class="fa fa-check"></i><b>3.25.1</b> Create directories in Unix</a></li>
<li class="chapter" data-level="3.25.2" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#create-an-rstudio-project"><i class="fa fa-check"></i><b>3.25.2</b> Create an RStudio project</a></li>
<li class="chapter" data-level="3.25.3" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#edit-some-r-scripts"><i class="fa fa-check"></i><b>3.25.3</b> Edit some R scripts</a></li>
<li class="chapter" data-level="3.25.4" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#create-some-more-directories-using-unix"><i class="fa fa-check"></i><b>3.25.4</b> Create some more directories using Unix</a></li>
<li class="chapter" data-level="3.25.5" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#add-a-readme-file"><i class="fa fa-check"></i><b>3.25.5</b> Add a README file</a></li>
<li class="chapter" data-level="3.25.6" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#initializing-a-git-directory"><i class="fa fa-check"></i><b>3.25.6</b> Initializing a Git directory</a></li>
<li class="chapter" data-level="3.25.7" data-path="3-productivity-tools.html"><a href="3-productivity-tools.html#add-commit-and-push-files-using-rstudio"><i class="fa fa-check"></i><b>3.25.7</b> Add, commit, and push files using RStudio</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="4" data-path="4-wrangling.html"><a href="4-wrangling.html"><i class="fa fa-check"></i><b>4</b> Data Wrangling</a><ul>
<li class="chapter" data-level="" data-path="4-wrangling.html"><a href="4-wrangling.html#needed-packages-1"><i class="fa fa-check"></i>Needed packages</a></li>
<li class="chapter" data-level="4.1" data-path="4-wrangling.html"><a href="4-wrangling.html#piping"><i class="fa fa-check"></i><b>4.1</b> The pipe operator: <code>%&gt;%</code></a></li>
<li class="chapter" data-level="4.2" data-path="4-wrangling.html"><a href="4-wrangling.html#filter"><i class="fa fa-check"></i><b>4.2</b> <code>filter</code> rows</a></li>
<li class="chapter" data-level="4.3" data-path="4-wrangling.html"><a href="4-wrangling.html#summarize"><i class="fa fa-check"></i><b>4.3</b> <code>summarize</code> variables</a></li>
<li class="chapter" data-level="4.4" data-path="4-wrangling.html"><a href="4-wrangling.html#groupby"><i class="fa fa-check"></i><b>4.4</b> <code>group_by</code> rows</a><ul>
<li class="chapter" data-level="4.4.1" data-path="4-wrangling.html"><a href="4-wrangling.html#grouping-by-more-than-one-variable"><i class="fa fa-check"></i><b>4.4.1</b> Grouping by more than one variable</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="4-wrangling.html"><a href="4-wrangling.html#mutate"><i class="fa fa-check"></i><b>4.5</b> <code>mutate</code> existing variables</a></li>
<li class="chapter" data-level="4.6" data-path="4-wrangling.html"><a href="4-wrangling.html#arrange"><i class="fa fa-check"></i><b>4.6</b> <code>arrange</code> and sort rows</a></li>
<li class="chapter" data-level="4.7" data-path="4-wrangling.html"><a href="4-wrangling.html#joins"><i class="fa fa-check"></i><b>4.7</b> <code>join</code> data frames</a><ul>
<li class="chapter" data-level="4.7.1" data-path="4-wrangling.html"><a href="4-wrangling.html#matching-key-variable-names"><i class="fa fa-check"></i><b>4.7.1</b> Matching “key” variable names</a></li>
<li class="chapter" data-level="4.7.2" data-path="4-wrangling.html"><a href="4-wrangling.html#diff-key"><i class="fa fa-check"></i><b>4.7.2</b> Different “key” variable names</a></li>
<li class="chapter" data-level="4.7.3" data-path="4-wrangling.html"><a href="4-wrangling.html#multiple-key-variables"><i class="fa fa-check"></i><b>4.7.3</b> Multiple “key” variables</a></li>
<li class="chapter" data-level="4.7.4" data-path="4-wrangling.html"><a href="4-wrangling.html#normal-forms"><i class="fa fa-check"></i><b>4.7.4</b> Normal forms</a></li>
</ul></li>
<li class="chapter" data-level="4.8" data-path="4-wrangling.html"><a href="4-wrangling.html#other-verbs"><i class="fa fa-check"></i><b>4.8</b> Other verbs</a><ul>
<li class="chapter" data-level="4.8.1" data-path="4-wrangling.html"><a href="4-wrangling.html#select"><i class="fa fa-check"></i><b>4.8.1</b> <code>select</code> variables</a></li>
<li class="chapter" data-level="4.8.2" data-path="4-wrangling.html"><a href="4-wrangling.html#rename"><i class="fa fa-check"></i><b>4.8.2</b> <code>rename</code> variables</a></li>
<li class="chapter" data-level="4.8.3" data-path="4-wrangling.html"><a href="4-wrangling.html#top_n-values-of-a-variable"><i class="fa fa-check"></i><b>4.8.3</b> <code>top_n</code> values of a variable</a></li>
</ul></li>
<li class="chapter" data-level="4.9" data-path="4-wrangling.html"><a href="4-wrangling.html#conclusion-2"><i class="fa fa-check"></i><b>4.9</b> Conclusion</a><ul>
<li class="chapter" data-level="4.9.1" data-path="4-wrangling.html"><a href="4-wrangling.html#summary-table-1"><i class="fa fa-check"></i><b>4.9.1</b> Summary table</a></li>
<li class="chapter" data-level="4.9.2" data-path="4-wrangling.html"><a href="4-wrangling.html#additional-resources-2"><i class="fa fa-check"></i><b>4.9.2</b> Additional resources</a></li>
<li class="chapter" data-level="4.9.3" data-path="4-wrangling.html"><a href="4-wrangling.html#whats-to-come-1"><i class="fa fa-check"></i><b>4.9.3</b> What’s to come?</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="5" data-path="5-tidy.html"><a href="5-tidy.html"><i class="fa fa-check"></i><b>5</b> Data Importing and “Tidy” Data</a><ul>
<li class="chapter" data-level="" data-path="5-tidy.html"><a href="5-tidy.html#needed-packages-2"><i class="fa fa-check"></i>Needed packages</a></li>
<li class="chapter" data-level="5.1" data-path="5-tidy.html"><a href="5-tidy.html#csv"><i class="fa fa-check"></i><b>5.1</b> Importing data</a><ul>
<li class="chapter" data-level="5.1.1" data-path="5-tidy.html"><a href="5-tidy.html#using-the-console"><i class="fa fa-check"></i><b>5.1.1</b> Using the console</a></li>
<li class="chapter" data-level="5.1.2" data-path="5-tidy.html"><a href="5-tidy.html#using-rstudios-interface"><i class="fa fa-check"></i><b>5.1.2</b> Using RStudio’s interface</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="5-tidy.html"><a href="5-tidy.html#tidy-data-ex"><i class="fa fa-check"></i><b>5.2</b> “Tidy” data</a><ul>
<li class="chapter" data-level="5.2.1" data-path="5-tidy.html"><a href="5-tidy.html#tidy-definition"><i class="fa fa-check"></i><b>5.2.1</b> Definition of “tidy” data</a></li>
<li class="chapter" data-level="5.2.2" data-path="5-tidy.html"><a href="5-tidy.html#converting-to-tidy-data"><i class="fa fa-check"></i><b>5.2.2</b> Converting to “tidy” data</a></li>
<li class="chapter" data-level="5.2.3" data-path="5-tidy.html"><a href="5-tidy.html#nycflights13-package-1"><i class="fa fa-check"></i><b>5.2.3</b> <code>nycflights13</code> package</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="5-tidy.html"><a href="5-tidy.html#case-study-tidy"><i class="fa fa-check"></i><b>5.3</b> Case study: Democracy in Guatemala</a></li>
<li class="chapter" data-level="5.4" data-path="5-tidy.html"><a href="5-tidy.html#tidyverse-package"><i class="fa fa-check"></i><b>5.4</b> <code>tidyverse</code> package</a></li>
<li class="chapter" data-level="5.5" data-path="5-tidy.html"><a href="5-tidy.html#conclusion-3"><i class="fa fa-check"></i><b>5.5</b> Conclusion</a><ul>
<li class="chapter" data-level="5.5.1" data-path="5-tidy.html"><a href="5-tidy.html#additional-resources-3"><i class="fa fa-check"></i><b>5.5.1</b> Additional resources</a></li>
<li class="chapter" data-level="5.5.2" data-path="5-tidy.html"><a href="5-tidy.html#whats-to-come-2"><i class="fa fa-check"></i><b>5.5.2</b> What’s to come?</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="6-bayess-theorem.html"><a href="6-bayess-theorem.html"><i class="fa fa-check"></i><b>6</b> Bayes’s Theorem</a><ul>
<li class="chapter" data-level="6.1" data-path="6-bayess-theorem.html"><a href="6-bayess-theorem.html#conditional-probability"><i class="fa fa-check"></i><b>6.1</b> Conditional probability</a></li>
<li class="chapter" data-level="6.2" data-path="6-bayess-theorem.html"><a href="6-bayess-theorem.html#conjoint-probability"><i class="fa fa-check"></i><b>6.2</b> Conjoint probability</a></li>
<li class="chapter" data-level="6.3" data-path="6-bayess-theorem.html"><a href="6-bayess-theorem.html#the-cookie-problem"><i class="fa fa-check"></i><b>6.3</b> The cookie problem</a></li>
<li class="chapter" data-level="6.4" data-path="6-bayess-theorem.html"><a href="6-bayess-theorem.html#bayess-theorem-1"><i class="fa fa-check"></i><b>6.4</b> Bayes’s Theorem</a></li>
<li class="chapter" data-level="6.5" data-path="6-bayess-theorem.html"><a href="6-bayess-theorem.html#the-diachronic-interpretation"><i class="fa fa-check"></i><b>6.5</b> The diachronic interpretation</a></li>
<li class="chapter" data-level="6.6" data-path="6-bayess-theorem.html"><a href="6-bayess-theorem.html#the-mm-problem"><i class="fa fa-check"></i><b>6.6</b> The M&amp;M problem</a></li>
<li class="chapter" data-level="6.7" data-path="6-bayess-theorem.html"><a href="6-bayess-theorem.html#the-monty-hall-problem"><i class="fa fa-check"></i><b>6.7</b> The Monty Hall problem</a></li>
<li class="chapter" data-level="6.8" data-path="6-bayess-theorem.html"><a href="6-bayess-theorem.html#discussion"><i class="fa fa-check"></i><b>6.8</b> Discussion</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="7-sampling.html"><a href="7-sampling.html"><i class="fa fa-check"></i><b>7</b> Sampling</a><ul>
<li class="chapter" data-level="" data-path="7-sampling.html"><a href="7-sampling.html#needed-packages-3"><i class="fa fa-check"></i>Needed packages</a></li>
<li class="chapter" data-level="7.1" data-path="7-sampling.html"><a href="7-sampling.html#sampling-activity"><i class="fa fa-check"></i><b>7.1</b> Sampling bowl activity</a><ul>
<li class="chapter" data-level="7.1.1" data-path="7-sampling.html"><a href="7-sampling.html#what-proportion-of-this-bowls-balls-are-red"><i class="fa fa-check"></i><b>7.1.1</b> What proportion of this bowl’s balls are red?</a></li>
<li class="chapter" data-level="7.1.2" data-path="7-sampling.html"><a href="7-sampling.html#using-the-shovel-once"><i class="fa fa-check"></i><b>7.1.2</b> Using the shovel once</a></li>
<li class="chapter" data-level="7.1.3" data-path="7-sampling.html"><a href="7-sampling.html#student-shovels"><i class="fa fa-check"></i><b>7.1.3</b> Using the shovel 33 times</a></li>
<li class="chapter" data-level="7.1.4" data-path="7-sampling.html"><a href="7-sampling.html#what-did-we-just-do"><i class="fa fa-check"></i><b>7.1.4</b> What did we just do?</a></li>
</ul></li>
<li class="chapter" data-level="7.2" data-path="7-sampling.html"><a href="7-sampling.html#sampling-simulation"><i class="fa fa-check"></i><b>7.2</b> Virtual sampling</a><ul>
<li class="chapter" data-level="7.2.1" data-path="7-sampling.html"><a href="7-sampling.html#using-the-virtual-shovel-once"><i class="fa fa-check"></i><b>7.2.1</b> Using the virtual shovel once</a></li>
<li class="chapter" data-level="7.2.2" data-path="7-sampling.html"><a href="7-sampling.html#using-the-virtual-shovel-33-times"><i class="fa fa-check"></i><b>7.2.2</b> Using the virtual shovel 33 times</a></li>
<li class="chapter" data-level="7.2.3" data-path="7-sampling.html"><a href="7-sampling.html#shovel-1000-times"><i class="fa fa-check"></i><b>7.2.3</b> Using the virtual shovel 1000 times</a></li>
<li class="chapter" data-level="7.2.4" data-path="7-sampling.html"><a href="7-sampling.html#different-shovels"><i class="fa fa-check"></i><b>7.2.4</b> Using different shovels</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="7-sampling.html"><a href="7-sampling.html#sampling-framework"><i class="fa fa-check"></i><b>7.3</b> Sampling framework</a><ul>
<li class="chapter" data-level="7.3.1" data-path="7-sampling.html"><a href="7-sampling.html#terminology-and-notation"><i class="fa fa-check"></i><b>7.3.1</b> Terminology and notation</a></li>
<li class="chapter" data-level="7.3.2" data-path="7-sampling.html"><a href="7-sampling.html#sampling-definitions"><i class="fa fa-check"></i><b>7.3.2</b> Statistical definitions</a></li>
<li class="chapter" data-level="7.3.3" data-path="7-sampling.html"><a href="7-sampling.html#moral-of-the-story"><i class="fa fa-check"></i><b>7.3.3</b> The moral of the story</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="7-sampling.html"><a href="7-sampling.html#sampling-case-study"><i class="fa fa-check"></i><b>7.4</b> Case study: Polls</a></li>
<li class="chapter" data-level="7.5" data-path="7-sampling.html"><a href="7-sampling.html#sampling-conclusion"><i class="fa fa-check"></i><b>7.5</b> Conclusion</a><ul>
<li class="chapter" data-level="7.5.1" data-path="7-sampling.html"><a href="7-sampling.html#sampling-conclusion-table"><i class="fa fa-check"></i><b>7.5.1</b> Sampling scenarios</a></li>
<li class="chapter" data-level="7.5.2" data-path="7-sampling.html"><a href="7-sampling.html#sampling-conclusion-central-limit-theorem"><i class="fa fa-check"></i><b>7.5.2</b> Central Limit Theorem</a></li>
<li class="chapter" data-level="7.5.3" data-path="7-sampling.html"><a href="7-sampling.html#additional-resources-4"><i class="fa fa-check"></i><b>7.5.3</b> Additional resources</a></li>
<li class="chapter" data-level="7.5.4" data-path="7-sampling.html"><a href="7-sampling.html#whats-to-come-4"><i class="fa fa-check"></i><b>7.5.4</b> What’s to come?</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html"><i class="fa fa-check"></i><b>8</b> Bootstrapping and Confidence Intervals</a><ul>
<li class="chapter" data-level="" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#needed-packages-4"><i class="fa fa-check"></i>Needed packages</a></li>
<li class="chapter" data-level="8.1" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#resampling-tactile"><i class="fa fa-check"></i><b>8.1</b> Pennies activity</a><ul>
<li class="chapter" data-level="8.1.1" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#what-is-the-average-year-on-us-pennies-in-2019"><i class="fa fa-check"></i><b>8.1.1</b> What is the average year on US pennies in 2019?</a></li>
<li class="chapter" data-level="8.1.2" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#resampling-once"><i class="fa fa-check"></i><b>8.1.2</b> Resampling once</a></li>
<li class="chapter" data-level="8.1.3" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#student-resamples"><i class="fa fa-check"></i><b>8.1.3</b> Resampling 35 times</a></li>
<li class="chapter" data-level="8.1.4" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#what-did-we-just-do-1"><i class="fa fa-check"></i><b>8.1.4</b> What did we just do?</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#resampling-simulation"><i class="fa fa-check"></i><b>8.2</b> Computer simulation of resampling</a><ul>
<li class="chapter" data-level="8.2.1" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#virtually-resampling-once"><i class="fa fa-check"></i><b>8.2.1</b> Virtually resampling once</a></li>
<li class="chapter" data-level="8.2.2" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#bootstrap-35-replicates"><i class="fa fa-check"></i><b>8.2.2</b> Virtually resampling 35 times</a></li>
<li class="chapter" data-level="8.2.3" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#bootstrap-1000-replicates"><i class="fa fa-check"></i><b>8.2.3</b> Virtually resampling 1000 times</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#ci-build-up"><i class="fa fa-check"></i><b>8.3</b> Understanding confidence intervals</a><ul>
<li class="chapter" data-level="8.3.1" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#percentile-method"><i class="fa fa-check"></i><b>8.3.1</b> Percentile method</a></li>
<li class="chapter" data-level="8.3.2" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#se-method"><i class="fa fa-check"></i><b>8.3.2</b> Standard error method</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#bootstrap-process"><i class="fa fa-check"></i><b>8.4</b> Constructing confidence intervals</a><ul>
<li class="chapter" data-level="8.4.1" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#original-workflow"><i class="fa fa-check"></i><b>8.4.1</b> Original workflow</a></li>
<li class="chapter" data-level="8.4.2" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#infer-workflow"><i class="fa fa-check"></i><b>8.4.2</b> <code>infer</code> package workflow</a></li>
<li class="chapter" data-level="8.4.3" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#percentile-method-infer"><i class="fa fa-check"></i><b>8.4.3</b> Percentile method with <code>infer</code></a></li>
<li class="chapter" data-level="8.4.4" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#infer-se"><i class="fa fa-check"></i><b>8.4.4</b> Standard error method with <code>infer</code></a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#one-prop-ci"><i class="fa fa-check"></i><b>8.5</b> Interpreting confidence intervals</a><ul>
<li class="chapter" data-level="8.5.1" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#ilyas-yohan"><i class="fa fa-check"></i><b>8.5.1</b> Did the net capture the fish?</a></li>
<li class="chapter" data-level="8.5.2" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#shorthand"><i class="fa fa-check"></i><b>8.5.2</b> Precise and shorthand interpretation</a></li>
<li class="chapter" data-level="8.5.3" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#ci-width"><i class="fa fa-check"></i><b>8.5.3</b> Width of confidence intervals</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#case-study-two-prop-ci"><i class="fa fa-check"></i><b>8.6</b> Case study: Is yawning contagious?</a><ul>
<li class="chapter" data-level="8.6.1" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#mythbusters-study-data"><i class="fa fa-check"></i><b>8.6.1</b> <em>Mythbusters</em> study data</a></li>
<li class="chapter" data-level="8.6.2" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#sampling-scenario"><i class="fa fa-check"></i><b>8.6.2</b> Sampling scenario</a></li>
<li class="chapter" data-level="8.6.3" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#ci-build"><i class="fa fa-check"></i><b>8.6.3</b> Constructing the confidence interval</a></li>
<li class="chapter" data-level="8.6.4" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#interpreting-the-confidence-interval"><i class="fa fa-check"></i><b>8.6.4</b> Interpreting the confidence interval</a></li>
</ul></li>
<li class="chapter" data-level="8.7" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#ci-conclusion"><i class="fa fa-check"></i><b>8.7</b> Conclusion</a><ul>
<li class="chapter" data-level="8.7.1" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#bootstrap-vs-sampling"><i class="fa fa-check"></i><b>8.7.1</b> Comparing bootstrap and sampling distributions</a></li>
<li class="chapter" data-level="8.7.2" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#theory-ci"><i class="fa fa-check"></i><b>8.7.2</b> Theory-based confidence intervals</a></li>
<li class="chapter" data-level="8.7.3" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#additional-resources-5"><i class="fa fa-check"></i><b>8.7.3</b> Additional resources</a></li>
<li class="chapter" data-level="8.7.4" data-path="8-confidence-intervals.html"><a href="8-confidence-intervals.html#whats-to-come-5"><i class="fa fa-check"></i><b>8.7.4</b> What’s to come?</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="9-regression.html"><a href="9-regression.html"><i class="fa fa-check"></i><b>9</b> Basic Regression</a><ul>
<li class="chapter" data-level="" data-path="9-regression.html"><a href="9-regression.html#needed-packages-5"><i class="fa fa-check"></i>Needed packages</a></li>
<li class="chapter" data-level="9.1" data-path="9-regression.html"><a href="9-regression.html#model1"><i class="fa fa-check"></i><b>9.1</b> One numerical explanatory variable</a><ul>
<li class="chapter" data-level="9.1.1" data-path="9-regression.html"><a href="9-regression.html#model1EDA"><i class="fa fa-check"></i><b>9.1.1</b> Exploratory data analysis</a></li>
<li class="chapter" data-level="9.1.2" data-path="9-regression.html"><a href="9-regression.html#model1table"><i class="fa fa-check"></i><b>9.1.2</b> Simple linear regression</a></li>
<li class="chapter" data-level="9.1.3" data-path="9-regression.html"><a href="9-regression.html#model1points"><i class="fa fa-check"></i><b>9.1.3</b> Observed/fitted values and residuals</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="9-regression.html"><a href="9-regression.html#model2"><i class="fa fa-check"></i><b>9.2</b> One categorical explanatory variable</a><ul>
<li class="chapter" data-level="9.2.1" data-path="9-regression.html"><a href="9-regression.html#model2EDA"><i class="fa fa-check"></i><b>9.2.1</b> Exploratory data analysis</a></li>
<li class="chapter" data-level="9.2.2" data-path="9-regression.html"><a href="9-regression.html#model2table"><i class="fa fa-check"></i><b>9.2.2</b> Linear regression</a></li>
<li class="chapter" data-level="9.2.3" data-path="9-regression.html"><a href="9-regression.html#model2points"><i class="fa fa-check"></i><b>9.2.3</b> Observed/fitted values and residuals</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="9-regression.html"><a href="9-regression.html#related-topics"><i class="fa fa-check"></i><b>9.3</b> Related topics</a><ul>
<li class="chapter" data-level="9.3.1" data-path="9-regression.html"><a href="9-regression.html#correlation-is-not-causation"><i class="fa fa-check"></i><b>9.3.1</b> Correlation is not necessarily causation</a></li>
<li class="chapter" data-level="9.3.2" data-path="9-regression.html"><a href="9-regression.html#leastsquares"><i class="fa fa-check"></i><b>9.3.2</b> Best-fitting line</a></li>
<li class="chapter" data-level="9.3.3" data-path="9-regression.html"><a href="9-regression.html#underthehood"><i class="fa fa-check"></i><b>9.3.3</b> <code>get_regression_x()</code> functions</a></li>
</ul></li>
<li class="chapter" data-level="9.4" data-path="9-regression.html"><a href="9-regression.html#conclusion-4"><i class="fa fa-check"></i><b>9.4</b> Conclusion</a><ul>
<li class="chapter" data-level="9.4.1" data-path="9-regression.html"><a href="9-regression.html#additional-resources-basic-regression"><i class="fa fa-check"></i><b>9.4.1</b> Additional resources</a></li>
<li class="chapter" data-level="9.4.2" data-path="9-regression.html"><a href="9-regression.html#whats-to-come-6"><i class="fa fa-check"></i><b>9.4.2</b> What’s to come?</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="10-multiple-regression.html"><a href="10-multiple-regression.html"><i class="fa fa-check"></i><b>10</b> Multiple Regression</a><ul>
<li class="chapter" data-level="" data-path="10-multiple-regression.html"><a href="10-multiple-regression.html#needed-packages-6"><i class="fa fa-check"></i>Needed packages</a></li>
<li class="chapter" data-level="10.1" data-path="10-multiple-regression.html"><a href="10-multiple-regression.html#model4"><i class="fa fa-check"></i><b>10.1</b> One numerical and one categorical explanatory variable</a><ul>
<li class="chapter" data-level="10.1.1" data-path="10-multiple-regression.html"><a href="10-multiple-regression.html#model4EDA"><i class="fa fa-check"></i><b>10.1.1</b> Exploratory data analysis</a></li>
<li class="chapter" data-level="10.1.2" data-path="10-multiple-regression.html"><a href="10-multiple-regression.html#model4interactiontable"><i class="fa fa-check"></i><b>10.1.2</b> Interaction model</a></li>
<li class="chapter" data-level="10.1.3" data-path="10-multiple-regression.html"><a href="10-multiple-regression.html#model4table"><i class="fa fa-check"></i><b>10.1.3</b> Parallel slopes model</a></li>
<li class="chapter" data-level="10.1.4" data-path="10-multiple-regression.html"><a href="10-multiple-regression.html#model4points"><i class="fa fa-check"></i><b>10.1.4</b> Observed/fitted values and residuals</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="10-multiple-regression.html"><a href="10-multiple-regression.html#model3"><i class="fa fa-check"></i><b>10.2</b> Two numerical explanatory variables</a><ul>
<li class="chapter" data-level="10.2.1" data-path="10-multiple-regression.html"><a href="10-multiple-regression.html#model3EDA"><i class="fa fa-check"></i><b>10.2.1</b> Exploratory data analysis</a></li>
<li class="chapter" data-level="10.2.2" data-path="10-multiple-regression.html"><a href="10-multiple-regression.html#model3table"><i class="fa fa-check"></i><b>10.2.2</b> Regression plane</a></li>
<li class="chapter" data-level="10.2.3" data-path="10-multiple-regression.html"><a href="10-multiple-regression.html#model3points"><i class="fa fa-check"></i><b>10.2.3</b> Observed/fitted values and residuals</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="10-multiple-regression.html"><a href="10-multiple-regression.html#related-topics-1"><i class="fa fa-check"></i><b>10.3</b> Related topics</a><ul>
<li class="chapter" data-level="10.3.1" data-path="10-multiple-regression.html"><a href="10-multiple-regression.html#model-selection"><i class="fa fa-check"></i><b>10.3.1</b> Model selection</a></li>
<li class="chapter" data-level="10.3.2" data-path="10-multiple-regression.html"><a href="10-multiple-regression.html#correlationcoefficient2"><i class="fa fa-check"></i><b>10.3.2</b> Correlation coefficient</a></li>
<li class="chapter" data-level="10.3.3" data-path="10-multiple-regression.html"><a href="10-multiple-regression.html#simpsonsparadox"><i class="fa fa-check"></i><b>10.3.3</b> Simpson’s Paradox</a></li>
</ul></li>
<li class="chapter" data-level="10.4" data-path="10-multiple-regression.html"><a href="10-multiple-regression.html#conclusion-5"><i class="fa fa-check"></i><b>10.4</b> Conclusion</a><ul>
<li class="chapter" data-level="10.4.1" data-path="10-multiple-regression.html"><a href="10-multiple-regression.html#additional-resources-6"><i class="fa fa-check"></i><b>10.4.1</b> Additional resources</a></li>
<li class="chapter" data-level="10.4.2" data-path="10-multiple-regression.html"><a href="10-multiple-regression.html#whats-to-come-7"><i class="fa fa-check"></i><b>10.4.2</b> What’s to come?</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>I Machine Learning</b></span></li>
<li class="chapter" data-level="11" data-path="11-introduction-to-machine-learning.html"><a href="11-introduction-to-machine-learning.html"><i class="fa fa-check"></i><b>11</b> Introduction to machine learning</a><ul>
<li class="chapter" data-level="11.1" data-path="11-introduction-to-machine-learning.html"><a href="11-introduction-to-machine-learning.html#notation"><i class="fa fa-check"></i><b>11.1</b> Notation</a></li>
<li class="chapter" data-level="11.2" data-path="11-introduction-to-machine-learning.html"><a href="11-introduction-to-machine-learning.html#an-example"><i class="fa fa-check"></i><b>11.2</b> An example</a></li>
<li class="chapter" data-level="11.3" data-path="11-introduction-to-machine-learning.html"><a href="11-introduction-to-machine-learning.html#exercises"><i class="fa fa-check"></i><b>11.3</b> Exercises</a></li>
<li class="chapter" data-level="11.4" data-path="11-introduction-to-machine-learning.html"><a href="11-introduction-to-machine-learning.html#evaluation-metrics"><i class="fa fa-check"></i><b>11.4</b> Evaluation metrics</a><ul>
<li class="chapter" data-level="11.4.1" data-path="11-introduction-to-machine-learning.html"><a href="11-introduction-to-machine-learning.html#training-and-test-sets"><i class="fa fa-check"></i><b>11.4.1</b> Training and test sets</a></li>
<li class="chapter" data-level="11.4.2" data-path="11-introduction-to-machine-learning.html"><a href="11-introduction-to-machine-learning.html#overall-accuracy"><i class="fa fa-check"></i><b>11.4.2</b> Overall accuracy</a></li>
<li class="chapter" data-level="11.4.3" data-path="11-introduction-to-machine-learning.html"><a href="11-introduction-to-machine-learning.html#the-confusion-matrix"><i class="fa fa-check"></i><b>11.4.3</b> The confusion matrix</a></li>
<li class="chapter" data-level="11.4.4" data-path="11-introduction-to-machine-learning.html"><a href="11-introduction-to-machine-learning.html#sensitivity-and-specificity"><i class="fa fa-check"></i><b>11.4.4</b> Sensitivity and specificity</a></li>
<li class="chapter" data-level="11.4.5" data-path="11-introduction-to-machine-learning.html"><a href="11-introduction-to-machine-learning.html#balanced-accuracy-and-f_1-score"><i class="fa fa-check"></i><b>11.4.5</b> Balanced accuracy and <span class="math inline">\(F_1\)</span> score</a></li>
<li class="chapter" data-level="11.4.6" data-path="11-introduction-to-machine-learning.html"><a href="11-introduction-to-machine-learning.html#prevalence-matters-in-practice"><i class="fa fa-check"></i><b>11.4.6</b> Prevalence matters in practice</a></li>
<li class="chapter" data-level="11.4.7" data-path="11-introduction-to-machine-learning.html"><a href="11-introduction-to-machine-learning.html#roc-and-precision-recall-curves"><i class="fa fa-check"></i><b>11.4.7</b> ROC and precision-recall curves</a></li>
<li class="chapter" data-level="11.4.8" data-path="11-introduction-to-machine-learning.html"><a href="11-introduction-to-machine-learning.html#loss-function"><i class="fa fa-check"></i><b>11.4.8</b> The loss function</a></li>
</ul></li>
<li class="chapter" data-level="11.5" data-path="11-introduction-to-machine-learning.html"><a href="11-introduction-to-machine-learning.html#exercises-1"><i class="fa fa-check"></i><b>11.5</b> Exercises</a></li>
<li class="chapter" data-level="11.6" data-path="11-introduction-to-machine-learning.html"><a href="11-introduction-to-machine-learning.html#conditional-probabilities-and-expectations"><i class="fa fa-check"></i><b>11.6</b> Conditional probabilities and expectations</a><ul>
<li class="chapter" data-level="11.6.1" data-path="11-introduction-to-machine-learning.html"><a href="11-introduction-to-machine-learning.html#conditional-probabilities"><i class="fa fa-check"></i><b>11.6.1</b> Conditional probabilities</a></li>
<li class="chapter" data-level="11.6.2" data-path="11-introduction-to-machine-learning.html"><a href="11-introduction-to-machine-learning.html#conditional-expectations"><i class="fa fa-check"></i><b>11.6.2</b> Conditional expectations</a></li>
<li class="chapter" data-level="11.6.3" data-path="11-introduction-to-machine-learning.html"><a href="11-introduction-to-machine-learning.html#conditional-expectation-minimizes-squared-loss-function"><i class="fa fa-check"></i><b>11.6.3</b> Conditional expectation minimizes squared loss function</a></li>
</ul></li>
<li class="chapter" data-level="11.7" data-path="11-introduction-to-machine-learning.html"><a href="11-introduction-to-machine-learning.html#exercises-2"><i class="fa fa-check"></i><b>11.7</b> Exercises</a></li>
<li class="chapter" data-level="11.8" data-path="11-introduction-to-machine-learning.html"><a href="11-introduction-to-machine-learning.html#two-or-seven"><i class="fa fa-check"></i><b>11.8</b> Case study: is it a 2 or a 7?</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="12-cross-validation.html"><a href="12-cross-validation.html"><i class="fa fa-check"></i><b>12</b> Cross validation</a><ul>
<li class="chapter" data-level="12.1" data-path="12-cross-validation.html"><a href="12-cross-validation.html#knn-cv-intro"><i class="fa fa-check"></i><b>12.1</b> Motivation with k-nearest neighbors</a><ul>
<li class="chapter" data-level="12.1.1" data-path="12-cross-validation.html"><a href="12-cross-validation.html#over-training"><i class="fa fa-check"></i><b>12.1.1</b> Over-training</a></li>
<li class="chapter" data-level="12.1.2" data-path="12-cross-validation.html"><a href="12-cross-validation.html#over-smoothing"><i class="fa fa-check"></i><b>12.1.2</b> Over-smoothing</a></li>
<li class="chapter" data-level="12.1.3" data-path="12-cross-validation.html"><a href="12-cross-validation.html#picking-the-k-in-knn"><i class="fa fa-check"></i><b>12.1.3</b> Picking the <span class="math inline">\(k\)</span> in kNN</a></li>
</ul></li>
<li class="chapter" data-level="12.2" data-path="12-cross-validation.html"><a href="12-cross-validation.html#mathematical-description-of-cross-validation"><i class="fa fa-check"></i><b>12.2</b> Mathematical description of cross validation</a></li>
<li class="chapter" data-level="12.3" data-path="12-cross-validation.html"><a href="12-cross-validation.html#k-fold-cross-validation"><i class="fa fa-check"></i><b>12.3</b> K-fold cross validation</a></li>
<li class="chapter" data-level="12.4" data-path="12-cross-validation.html"><a href="12-cross-validation.html#exercises-3"><i class="fa fa-check"></i><b>12.4</b> Exercises</a></li>
<li class="chapter" data-level="12.5" data-path="12-cross-validation.html"><a href="12-cross-validation.html#bootstrap"><i class="fa fa-check"></i><b>12.5</b> Bootstrap</a></li>
<li class="chapter" data-level="12.6" data-path="12-cross-validation.html"><a href="12-cross-validation.html#exercises-4"><i class="fa fa-check"></i><b>12.6</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="13-examples-of-algorithms.html"><a href="13-examples-of-algorithms.html"><i class="fa fa-check"></i><b>13</b> Examples of algorithms</a><ul>
<li class="chapter" data-level="13.1" data-path="13-examples-of-algorithms.html"><a href="13-examples-of-algorithms.html#linear-regression"><i class="fa fa-check"></i><b>13.1</b> Linear regression</a><ul>
<li class="chapter" data-level="13.1.1" data-path="13-examples-of-algorithms.html"><a href="13-examples-of-algorithms.html#the-predict-function"><i class="fa fa-check"></i><b>13.1.1</b> The <code>predict</code> function</a></li>
</ul></li>
<li class="chapter" data-level="13.2" data-path="13-examples-of-algorithms.html"><a href="13-examples-of-algorithms.html#exercises-5"><i class="fa fa-check"></i><b>13.2</b> Exercises</a></li>
<li class="chapter" data-level="13.3" data-path="13-examples-of-algorithms.html"><a href="13-examples-of-algorithms.html#logistic-regression"><i class="fa fa-check"></i><b>13.3</b> Logistic regression</a><ul>
<li class="chapter" data-level="13.3.1" data-path="13-examples-of-algorithms.html"><a href="13-examples-of-algorithms.html#generalized-linear-models"><i class="fa fa-check"></i><b>13.3.1</b> Generalized linear models</a></li>
<li class="chapter" data-level="13.3.2" data-path="13-examples-of-algorithms.html"><a href="13-examples-of-algorithms.html#logistic-regression-with-more-than-one-predictor"><i class="fa fa-check"></i><b>13.3.2</b> Logistic regression with more than one predictor</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="13-examples-of-algorithms.html"><a href="13-examples-of-algorithms.html#exercises-6"><i class="fa fa-check"></i><b>13.4</b> Exercises</a></li>
<li class="chapter" data-level="13.5" data-path="13-examples-of-algorithms.html"><a href="13-examples-of-algorithms.html#k-nearest-neighbors"><i class="fa fa-check"></i><b>13.5</b> k-nearest neighbors</a></li>
<li class="chapter" data-level="13.6" data-path="13-examples-of-algorithms.html"><a href="13-examples-of-algorithms.html#exercises-7"><i class="fa fa-check"></i><b>13.6</b> Exercises</a></li>
<li class="chapter" data-level="13.7" data-path="13-examples-of-algorithms.html"><a href="13-examples-of-algorithms.html#generative-models"><i class="fa fa-check"></i><b>13.7</b> Generative models</a><ul>
<li class="chapter" data-level="13.7.1" data-path="13-examples-of-algorithms.html"><a href="13-examples-of-algorithms.html#naive-bayes"><i class="fa fa-check"></i><b>13.7.1</b> Naive Bayes</a></li>
<li class="chapter" data-level="13.7.2" data-path="13-examples-of-algorithms.html"><a href="13-examples-of-algorithms.html#controlling-prevalence"><i class="fa fa-check"></i><b>13.7.2</b> Controlling prevalence</a></li>
<li class="chapter" data-level="13.7.3" data-path="13-examples-of-algorithms.html"><a href="13-examples-of-algorithms.html#quadratic-discriminant-analysis"><i class="fa fa-check"></i><b>13.7.3</b> Quadratic discriminant analysis</a></li>
<li class="chapter" data-level="13.7.4" data-path="13-examples-of-algorithms.html"><a href="13-examples-of-algorithms.html#linear-discriminant-analysis"><i class="fa fa-check"></i><b>13.7.4</b> Linear discriminant analysis</a></li>
<li class="chapter" data-level="13.7.5" data-path="13-examples-of-algorithms.html"><a href="13-examples-of-algorithms.html#connection-to-distance"><i class="fa fa-check"></i><b>13.7.5</b> Connection to distance</a></li>
</ul></li>
<li class="chapter" data-level="13.8" data-path="13-examples-of-algorithms.html"><a href="13-examples-of-algorithms.html#case-study-more-than-three-classes"><i class="fa fa-check"></i><b>13.8</b> Case study: more than three classes</a></li>
<li class="chapter" data-level="13.9" data-path="13-examples-of-algorithms.html"><a href="13-examples-of-algorithms.html#exercises-8"><i class="fa fa-check"></i><b>13.9</b> Exercises</a></li>
<li class="chapter" data-level="13.10" data-path="13-examples-of-algorithms.html"><a href="13-examples-of-algorithms.html#classification-and-regression-trees-cart"><i class="fa fa-check"></i><b>13.10</b> Classification and regression trees (CART)</a><ul>
<li class="chapter" data-level="13.10.1" data-path="13-examples-of-algorithms.html"><a href="13-examples-of-algorithms.html#the-curse-of-dimensionality"><i class="fa fa-check"></i><b>13.10.1</b> The curse of dimensionality</a></li>
<li class="chapter" data-level="13.10.2" data-path="13-examples-of-algorithms.html"><a href="13-examples-of-algorithms.html#cart-motivation"><i class="fa fa-check"></i><b>13.10.2</b> CART motivation</a></li>
<li class="chapter" data-level="13.10.3" data-path="13-examples-of-algorithms.html"><a href="13-examples-of-algorithms.html#regression-trees"><i class="fa fa-check"></i><b>13.10.3</b> Regression trees</a></li>
<li class="chapter" data-level="13.10.4" data-path="13-examples-of-algorithms.html"><a href="13-examples-of-algorithms.html#classification-decision-trees"><i class="fa fa-check"></i><b>13.10.4</b> Classification (decision) trees</a></li>
</ul></li>
<li class="chapter" data-level="13.11" data-path="13-examples-of-algorithms.html"><a href="13-examples-of-algorithms.html#random-forests"><i class="fa fa-check"></i><b>13.11</b> Random forests</a></li>
<li class="chapter" data-level="13.12" data-path="13-examples-of-algorithms.html"><a href="13-examples-of-algorithms.html#exercises-9"><i class="fa fa-check"></i><b>13.12</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="14-thinking-with-data.html"><a href="14-thinking-with-data.html"><i class="fa fa-check"></i><b>14</b> Tell Your Story with Data</a><ul>
<li class="chapter" data-level="14.1" data-path="14-thinking-with-data.html"><a href="14-thinking-with-data.html#review"><i class="fa fa-check"></i><b>14.1</b> Review</a><ul>
<li class="chapter" data-level="" data-path="14-thinking-with-data.html"><a href="14-thinking-with-data.html#needed-packages-7"><i class="fa fa-check"></i>Needed packages</a></li>
</ul></li>
<li class="chapter" data-level="14.2" data-path="14-thinking-with-data.html"><a href="14-thinking-with-data.html#seattle-house-prices"><i class="fa fa-check"></i><b>14.2</b> Case study: Seattle house prices</a><ul>
<li class="chapter" data-level="14.2.1" data-path="14-thinking-with-data.html"><a href="14-thinking-with-data.html#house-prices-EDA-I"><i class="fa fa-check"></i><b>14.2.1</b> Exploratory data analysis: Part I</a></li>
<li class="chapter" data-level="14.2.2" data-path="14-thinking-with-data.html"><a href="14-thinking-with-data.html#house-prices-EDA-II"><i class="fa fa-check"></i><b>14.2.2</b> Exploratory data analysis: Part II</a></li>
<li class="chapter" data-level="14.2.3" data-path="14-thinking-with-data.html"><a href="14-thinking-with-data.html#house-prices-regression"><i class="fa fa-check"></i><b>14.2.3</b> Regression modeling</a></li>
<li class="chapter" data-level="14.2.4" data-path="14-thinking-with-data.html"><a href="14-thinking-with-data.html#house-prices-making-predictions"><i class="fa fa-check"></i><b>14.2.4</b> Making predictions</a></li>
</ul></li>
<li class="chapter" data-level="14.3" data-path="14-thinking-with-data.html"><a href="14-thinking-with-data.html#data-journalism"><i class="fa fa-check"></i><b>14.3</b> Case study: Effective data storytelling</a><ul>
<li class="chapter" data-level="14.3.1" data-path="14-thinking-with-data.html"><a href="14-thinking-with-data.html#bechdel-test-for-hollywood-gender-representation"><i class="fa fa-check"></i><b>14.3.1</b> Bechdel test for Hollywood gender representation</a></li>
<li class="chapter" data-level="14.3.2" data-path="14-thinking-with-data.html"><a href="14-thinking-with-data.html#us-births-in-1999"><i class="fa fa-check"></i><b>14.3.2</b> US Births in 1999</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="14-thinking-with-data.html"><a href="14-thinking-with-data.html#concluding-remarks"><i class="fa fa-check"></i>Concluding remarks</a></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li class="chapter" data-level="A" data-path="A-appendixA.html"><a href="A-appendixA.html"><i class="fa fa-check"></i><b>A</b> Statistical Background</a><ul>
<li class="chapter" data-level="A.1" data-path="A-appendixA.html"><a href="A-appendixA.html#appendix-stat-terms"><i class="fa fa-check"></i><b>A.1</b> Basic statistical terms</a><ul>
<li class="chapter" data-level="A.1.1" data-path="A-appendixA.html"><a href="A-appendixA.html#mean"><i class="fa fa-check"></i><b>A.1.1</b> Mean</a></li>
<li class="chapter" data-level="A.1.2" data-path="A-appendixA.html"><a href="A-appendixA.html#median"><i class="fa fa-check"></i><b>A.1.2</b> Median</a></li>
<li class="chapter" data-level="A.1.3" data-path="A-appendixA.html"><a href="A-appendixA.html#standard-deviation"><i class="fa fa-check"></i><b>A.1.3</b> Standard deviation</a></li>
<li class="chapter" data-level="A.1.4" data-path="A-appendixA.html"><a href="A-appendixA.html#five-number-summary"><i class="fa fa-check"></i><b>A.1.4</b> Five-number summary</a></li>
<li class="chapter" data-level="A.1.5" data-path="A-appendixA.html"><a href="A-appendixA.html#distribution"><i class="fa fa-check"></i><b>A.1.5</b> Distribution</a></li>
<li class="chapter" data-level="A.1.6" data-path="A-appendixA.html"><a href="A-appendixA.html#outliers"><i class="fa fa-check"></i><b>A.1.6</b> Outliers</a></li>
</ul></li>
<li class="chapter" data-level="A.2" data-path="A-appendixA.html"><a href="A-appendixA.html#appendix-normal-curve"><i class="fa fa-check"></i><b>A.2</b> Normal distribution</a></li>
<li class="chapter" data-level="A.3" data-path="A-appendixA.html"><a href="A-appendixA.html#appendix-log10-transformations"><i class="fa fa-check"></i><b>A.3</b> log10 transformations</a></li>
</ul></li>
<li class="chapter" data-level="B" data-path="B-appendixC.html"><a href="B-appendixC.html"><i class="fa fa-check"></i><b>B</b> Reach for the Stars</a><ul>
<li class="chapter" data-level="" data-path="B-appendixC.html"><a href="B-appendixC.html#needed-packages-8"><i class="fa fa-check"></i>Needed packages</a></li>
<li class="chapter" data-level="B.1" data-path="B-appendixC.html"><a href="B-appendixC.html#sorted-barplots"><i class="fa fa-check"></i><b>B.1</b> Sorted barplots</a></li>
<li class="chapter" data-level="B.2" data-path="B-appendixC.html"><a href="B-appendixC.html#interactive-graphics"><i class="fa fa-check"></i><b>B.2</b> Interactive graphics</a><ul>
<li class="chapter" data-level="B.2.1" data-path="B-appendixC.html"><a href="B-appendixC.html#interactive-linegraphs"><i class="fa fa-check"></i><b>B.2.1</b> Interactive linegraphs</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="C" data-path="C-appendixE.html"><a href="C-appendixE.html"><i class="fa fa-check"></i><b>C</b> Versions of R Packages Used</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Preceptor’s Primer for Bayesian Data Science</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="examples-of-algorithms" class="section level1">
<h1><span class="header-section-number">Chapter 13</span> Examples of algorithms</h1>
<p>There are dozens of machine learning algorithms. Here we provide a few examples spanning rather different approaches. Throughout the chapter we will be using the two predictor digits data introduced in Section <a href="11-introduction-to-machine-learning.html#two-or-seven">11.8</a> to demonstrate how the algorithms work.</p>
<div class="sourceCode" id="cb550"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb550-1" data-line-number="1"><span class="kw">library</span>(tidyverse)</a>
<a class="sourceLine" id="cb550-2" data-line-number="2"><span class="kw">library</span>(dslabs)</a>
<a class="sourceLine" id="cb550-3" data-line-number="3"><span class="kw">library</span>(caret)</a>
<a class="sourceLine" id="cb550-4" data-line-number="4"><span class="kw">data</span>(<span class="st">&quot;mnist_27&quot;</span>)</a></code></pre></div>
<div id="linear-regression" class="section level2">
<h2><span class="header-section-number">13.1</span> Linear regression</h2>
<p>Linear regression can be considered a machine learning algorithm. In Section <a href="11-introduction-to-machine-learning.html#two-or-seven">11.8</a> we demonstrated how linear regression can be too rigid to be useful. This is generally true, but for some challenges it works rather well. It also serves as a baseline approach: if you can’t beat it with a more complex approach, you probably want to stick to linear regression. To quickly make the connection between regression and machine learning, we will reformulate Galton’s study with heights, a continuous outcome.</p>
<div class="sourceCode" id="cb551"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb551-1" data-line-number="1"><span class="kw">library</span>(HistData)</a>
<a class="sourceLine" id="cb551-2" data-line-number="2"></a>
<a class="sourceLine" id="cb551-3" data-line-number="3"><span class="kw">set.seed</span>(<span class="dv">1983</span>)</a>
<a class="sourceLine" id="cb551-4" data-line-number="4">galton_heights &lt;-<span class="st"> </span>GaltonFamilies <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb551-5" data-line-number="5"><span class="st">  </span><span class="kw">filter</span>(gender <span class="op">==</span><span class="st"> &quot;male&quot;</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb551-6" data-line-number="6"><span class="st">  </span><span class="kw">group_by</span>(family) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb551-7" data-line-number="7"><span class="st">  </span><span class="kw">sample_n</span>(<span class="dv">1</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb551-8" data-line-number="8"><span class="st">  </span><span class="kw">ungroup</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb551-9" data-line-number="9"><span class="st">  </span><span class="kw">select</span>(father, childHeight) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb551-10" data-line-number="10"><span class="st">  </span><span class="kw">rename</span>(<span class="dt">son =</span> childHeight)</a></code></pre></div>
<p>Suppose you are tasked with building a machine learning algorithm that predicts the son’s height <span class="math inline">\(Y\)</span> using the father’s height <span class="math inline">\(X\)</span>. Let’s generate testing and training sets:</p>
<div class="sourceCode" id="cb552"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb552-1" data-line-number="1">y &lt;-<span class="st"> </span>galton_heights<span class="op">$</span>son</a>
<a class="sourceLine" id="cb552-2" data-line-number="2">test_index &lt;-<span class="st"> </span><span class="kw">createDataPartition</span>(y, <span class="dt">times =</span> <span class="dv">1</span>, <span class="dt">p =</span> <span class="fl">0.5</span>, <span class="dt">list =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb552-3" data-line-number="3"></a>
<a class="sourceLine" id="cb552-4" data-line-number="4">train_set &lt;-<span class="st"> </span>galton_heights <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">slice</span>(<span class="op">-</span>test_index)</a>
<a class="sourceLine" id="cb552-5" data-line-number="5">test_set &lt;-<span class="st"> </span>galton_heights <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">slice</span>(test_index)</a></code></pre></div>
<p>In this case, if we were just ignoring the father’s height and guessing the son’s height, we would guess the average height of sons.</p>
<div class="sourceCode" id="cb553"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb553-1" data-line-number="1">m &lt;-<span class="st"> </span><span class="kw">mean</span>(train_set<span class="op">$</span>son)</a>
<a class="sourceLine" id="cb553-2" data-line-number="2">m</a></code></pre></div>
<pre><code>[1] 69.2</code></pre>
<p>Our squared loss is:</p>
<div class="sourceCode" id="cb555"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb555-1" data-line-number="1"><span class="kw">mean</span>((m <span class="op">-</span><span class="st"> </span>test_set<span class="op">$</span>son)<span class="op">^</span><span class="dv">2</span>)</a></code></pre></div>
<pre><code>[1] 7.65</code></pre>
<p>Can we do better? In the regression chapter, we learned that if the pair <span class="math inline">\((X,Y)\)</span> follow a bivariate normal distribution, the conditional expectation (what we want to estimate) is equivalent to the regression line:</p>
<p><span class="math display">\[
f(x) = \mbox{E}( Y  \mid  X= x ) = \beta_0 + \beta_1 x
\]</span></p>
<p>In Section <a href="#lse"><strong>??</strong></a> we introduced least squares as a method for estimating the slope <span class="math inline">\(\beta_0\)</span> and intercept <span class="math inline">\(\beta_1\)</span>:</p>
<div class="sourceCode" id="cb557"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb557-1" data-line-number="1">fit &lt;-<span class="st"> </span><span class="kw">lm</span>(son <span class="op">~</span><span class="st"> </span>father, <span class="dt">data =</span> train_set)</a>
<a class="sourceLine" id="cb557-2" data-line-number="2">fit<span class="op">$</span>coef</a></code></pre></div>
<pre><code>(Intercept)      father 
     35.976       0.482 </code></pre>
<p>This gives us an estimate of the conditional expectation:</p>
<p><span class="math display">\[ \hat{f}(x) = 52 + 0.25 x \]</span></p>
<p>We can see that this does indeed provide an improvement over our guessing approach.</p>
<div class="sourceCode" id="cb559"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb559-1" data-line-number="1">y_hat &lt;-<span class="st"> </span>fit<span class="op">$</span>coef[<span class="dv">1</span>] <span class="op">+</span><span class="st"> </span>fit<span class="op">$</span>coef[<span class="dv">2</span>]<span class="op">*</span>test_set<span class="op">$</span>father</a>
<a class="sourceLine" id="cb559-2" data-line-number="2"><span class="kw">mean</span>((y_hat <span class="op">-</span><span class="st"> </span>test_set<span class="op">$</span>son)<span class="op">^</span><span class="dv">2</span>)</a></code></pre></div>
<pre><code>[1] 6.47</code></pre>
<div id="the-predict-function" class="section level3">
<h3><span class="header-section-number">13.1.1</span> The <code>predict</code> function</h3>
<p>The <code>predict</code> function is very useful for machine learning applications. This function takes a fitted object from functions such as <code>lm</code> or <code>glm</code> (we learn about <code>glm</code> soon) and a data frame with the new predictors for which to predict. So in our current example, we would use <code>predict</code> like this:</p>
<div class="sourceCode" id="cb561"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb561-1" data-line-number="1">y_hat &lt;-<span class="st"> </span><span class="kw">predict</span>(fit, test_set)</a></code></pre></div>
<p>Using <code>predict</code>, we can get the same results as we did previously:</p>
<div class="sourceCode" id="cb562"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb562-1" data-line-number="1">y_hat &lt;-<span class="st"> </span><span class="kw">predict</span>(fit, test_set)</a>
<a class="sourceLine" id="cb562-2" data-line-number="2"><span class="kw">mean</span>((y_hat <span class="op">-</span><span class="st"> </span>test_set<span class="op">$</span>son)<span class="op">^</span><span class="dv">2</span>)</a></code></pre></div>
<pre><code>[1] 6.47</code></pre>
<p><code>predict</code> does not always return objects of the same types; it depends on what type of object is sent to it. To learn about the specifics, you need to look at the help file specific for the type of fit object that is being used. The <code>predict</code> is actually a special type of function in R (called a <em>generic function</em>) that calls other functions depending on what kind of object it receives. So if <code>predict</code> receives an object coming out of the <code>lm</code> function, it will call <code>predict.lm</code>. If it receives an object coming out of <code>glm</code>, it calls <code>predict.glm</code>. These two functions are similar but different. You can learn more about the differences by reading the help files:</p>
<div class="sourceCode" id="cb564"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb564-1" data-line-number="1">?predict.lm</a>
<a class="sourceLine" id="cb564-2" data-line-number="2">?predict.glm</a></code></pre></div>
<p>There are many other versions of <code>predict</code> and many machine learning algorithms have a <code>predict</code> function.</p>
</div>
</div>
<div id="exercises-5" class="section level2">
<h2><span class="header-section-number">13.2</span> Exercises</h2>
<p>1. Create a dataset using the following code.</p>
<div class="sourceCode" id="cb565"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb565-1" data-line-number="1">n &lt;-<span class="st"> </span><span class="dv">100</span></a>
<a class="sourceLine" id="cb565-2" data-line-number="2">Sigma &lt;-<span class="st"> </span><span class="dv">9</span><span class="op">*</span><span class="kw">matrix</span>(<span class="kw">c</span>(<span class="fl">1.0</span>, <span class="fl">0.5</span>, <span class="fl">0.5</span>, <span class="fl">1.0</span>), <span class="dv">2</span>, <span class="dv">2</span>)</a>
<a class="sourceLine" id="cb565-3" data-line-number="3">dat &lt;-<span class="st"> </span>MASS<span class="op">::</span><span class="kw">mvrnorm</span>(<span class="dt">n =</span> <span class="dv">100</span>, <span class="kw">c</span>(<span class="dv">69</span>, <span class="dv">69</span>), Sigma) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb565-4" data-line-number="4"><span class="st">  </span><span class="kw">data.frame</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">setNames</span>(<span class="kw">c</span>(<span class="st">&quot;x&quot;</span>, <span class="st">&quot;y&quot;</span>))</a></code></pre></div>
<p>Use the <strong>caret</strong> package to partition into a test and training set of equal size. Train a linear model and report the RMSE. Repeat this exercise 100 times and make a histogram of the RMSEs and report the average and standard deviation. Hint: adapt the code shown earlier like this:</p>
<div class="sourceCode" id="cb566"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb566-1" data-line-number="1">y &lt;-<span class="st"> </span>dat<span class="op">$</span>y</a>
<a class="sourceLine" id="cb566-2" data-line-number="2">test_index &lt;-<span class="st"> </span><span class="kw">createDataPartition</span>(y, <span class="dt">times =</span> <span class="dv">1</span>, <span class="dt">p =</span> <span class="fl">0.5</span>, <span class="dt">list =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb566-3" data-line-number="3">train_set &lt;-<span class="st"> </span>dat <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">slice</span>(<span class="op">-</span>test_index)</a>
<a class="sourceLine" id="cb566-4" data-line-number="4">test_set &lt;-<span class="st"> </span>dat <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">slice</span>(test_index)</a>
<a class="sourceLine" id="cb566-5" data-line-number="5">fit &lt;-<span class="st"> </span><span class="kw">lm</span>(y <span class="op">~</span><span class="st"> </span>x, <span class="dt">data =</span> train_set)</a>
<a class="sourceLine" id="cb566-6" data-line-number="6">y_hat &lt;-<span class="st"> </span>fit<span class="op">$</span>coef[<span class="dv">1</span>] <span class="op">+</span><span class="st"> </span>fit<span class="op">$</span>coef[<span class="dv">2</span>]<span class="op">*</span>test_set<span class="op">$</span>x</a>
<a class="sourceLine" id="cb566-7" data-line-number="7"><span class="kw">mean</span>((y_hat <span class="op">-</span><span class="st"> </span>test_set<span class="op">$</span>y)<span class="op">^</span><span class="dv">2</span>)</a></code></pre></div>
<p>and put it inside a call to <code>replicate</code>.</p>
<p>2. Now we will repeat the above but using larger datasets. Repeat exercise 1 but for datasets with <code>n &lt;- c(100, 500, 1000, 5000, 10000)</code>. Save the average and standard deviation of RMSE from the 100 repetitions. Hint: use the <code>sapply</code> or <code>map</code> functions.</p>
<p>3. Describe what you observe with the RMSE as the size of the dataset becomes larger.</p>
<ol style="list-style-type: lower-alpha">
<li>On average, the RMSE does not change much as <code>n</code> gets larger, while the variability of RMSE does decrease.</li>
<li>Because of the law of large numbers, the RMSE decreases: more data, more precise estimates.</li>
<li><code>n = 10000</code> is not sufficiently large. To see a decrease in RMSE, we need to make it larger.</li>
<li>The RMSE is not a random variable.</li>
</ol>
<p>4. Now repeat exercise 1, but this time make the correlation between <code>x</code> and <code>y</code> larger by changing <code>Sigma</code> like this:</p>
<div class="sourceCode" id="cb567"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb567-1" data-line-number="1">n &lt;-<span class="st"> </span><span class="dv">100</span></a>
<a class="sourceLine" id="cb567-2" data-line-number="2">Sigma &lt;-<span class="st"> </span><span class="dv">9</span><span class="op">*</span><span class="kw">matrix</span>(<span class="kw">c</span>(<span class="dv">1</span>, <span class="fl">0.95</span>, <span class="fl">0.95</span>, <span class="dv">1</span>), <span class="dv">2</span>, <span class="dv">2</span>)</a>
<a class="sourceLine" id="cb567-3" data-line-number="3">dat &lt;-<span class="st"> </span>MASS<span class="op">::</span><span class="kw">mvrnorm</span>(<span class="dt">n =</span> <span class="dv">100</span>, <span class="kw">c</span>(<span class="dv">69</span>, <span class="dv">69</span>), Sigma) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb567-4" data-line-number="4"><span class="st">  </span><span class="kw">data.frame</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">setNames</span>(<span class="kw">c</span>(<span class="st">&quot;x&quot;</span>, <span class="st">&quot;y&quot;</span>))</a></code></pre></div>
<p>Repeat the exercise and note what happens to the RMSE now.</p>
<p>5. Which of the following best explains why the RMSE in exercise 4 is so much lower than exercise 1.</p>
<ol style="list-style-type: lower-alpha">
<li>It is just luck. If we do it again, it will be larger.</li>
<li>The Central Limit Theorem tells us the RMSE is normal.</li>
<li>When we increase the correlation between <code>x</code> and <code>y</code>, <code>x</code> has more predictive power and thus provides a better estimate of <code>y</code>. This correlation has a much bigger effect on RMSE than <code>n</code>. Large <code>n</code> simply provide us more precise estimates of the linear model coefficients.</li>
<li>These are both examples of regression, so the RMSE has to be the same.</li>
</ol>
<p>6. Create a dataset using the following code:</p>
<div class="sourceCode" id="cb568"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb568-1" data-line-number="1">n &lt;-<span class="st"> </span><span class="dv">1000</span></a>
<a class="sourceLine" id="cb568-2" data-line-number="2">Sigma &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">3</span><span class="op">/</span><span class="dv">4</span>, <span class="dv">3</span><span class="op">/</span><span class="dv">4</span>, <span class="dv">3</span><span class="op">/</span><span class="dv">4</span>, <span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">3</span><span class="op">/</span><span class="dv">4</span>, <span class="dv">0</span>, <span class="dv">1</span>), <span class="dv">3</span>, <span class="dv">3</span>)</a>
<a class="sourceLine" id="cb568-3" data-line-number="3">dat &lt;-<span class="st"> </span>MASS<span class="op">::</span><span class="kw">mvrnorm</span>(<span class="dt">n =</span> <span class="dv">100</span>, <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>), Sigma) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb568-4" data-line-number="4"><span class="st">  </span><span class="kw">data.frame</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">setNames</span>(<span class="kw">c</span>(<span class="st">&quot;y&quot;</span>, <span class="st">&quot;x_1&quot;</span>, <span class="st">&quot;x_2&quot;</span>))</a></code></pre></div>
<p>Note that <code>y</code> is correlated with both <code>x_1</code> and <code>x_2</code>, but the two predictors are independent of each other.</p>
<div class="sourceCode" id="cb569"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb569-1" data-line-number="1"><span class="kw">cor</span>(dat)</a></code></pre></div>
<p>Use the <strong>caret</strong> package to partition into a test and training set of equal size. Compare the RMSE when using just <code>x_1</code>, just <code>x_2</code>, and both <code>x_1</code> and <code>x_2</code>. Train a linear model and report the RMSE.</p>
<p>7. Repeat exercise 6 but now create an example in which <code>x_1</code> and <code>x_2</code> are highly correlated:</p>
<div class="sourceCode" id="cb570"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb570-1" data-line-number="1">n &lt;-<span class="st"> </span><span class="dv">1000</span></a>
<a class="sourceLine" id="cb570-2" data-line-number="2">Sigma &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">c</span>(<span class="fl">1.0</span>, <span class="fl">0.75</span>, <span class="fl">0.75</span>, <span class="fl">0.75</span>, <span class="fl">1.0</span>, <span class="fl">0.95</span>, <span class="fl">0.75</span>, <span class="fl">0.95</span>, <span class="fl">1.0</span>), <span class="dv">3</span>, <span class="dv">3</span>)</a>
<a class="sourceLine" id="cb570-3" data-line-number="3">dat &lt;-<span class="st"> </span>MASS<span class="op">::</span><span class="kw">mvrnorm</span>(<span class="dt">n =</span> <span class="dv">100</span>, <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>), Sigma) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb570-4" data-line-number="4"><span class="st">  </span><span class="kw">data.frame</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">setNames</span>(<span class="kw">c</span>(<span class="st">&quot;y&quot;</span>, <span class="st">&quot;x_1&quot;</span>, <span class="st">&quot;x_2&quot;</span>))</a></code></pre></div>
<p>Use the <strong>caret</strong> package to partition into a test and training set of equal size. Compare the RMSE when using just <code>x_1</code>, just <code>x_2</code>, and both <code>x_1</code> and <code>x_2</code> Train a linear model and report the RMSE.</p>
<p>8. Compare the results in 6 and 7 and choose the statement you agree with:</p>
<ol style="list-style-type: lower-alpha">
<li>Adding extra predictors can improve RMSE substantially, but not when they are highly correlated with another predictor.</li>
<li>Adding extra predictors improves predictions equally in both exercises.</li>
<li>Adding extra predictors results in over fitting.</li>
<li>Unless we include all predictors, we have no predicting power.</li>
</ol>
</div>
<div id="logistic-regression" class="section level2">
<h2><span class="header-section-number">13.3</span> Logistic regression</h2>
<p>The regression approach can be extended to categorical data. In this section we first illustrate how, for binary data, one can simply assign numeric values of 0 and 1 to the outcomes <span class="math inline">\(y\)</span>, and apply regression as if the data were continuous. We will then point out a limitation with this approach and introduce <em>logistic regression</em> as a solution. Logistic regression is a specific case of a set of <em>generalized linear models</em>. To illustrate logistic regression, we will apply it to our previous predicting sex example:</p>
<p>If we define the outcome <span class="math inline">\(Y\)</span> as 1 for females and 0 for males, and <span class="math inline">\(X\)</span> as the height, we are interested in the conditional probability:</p>
<p><span class="math display">\[
\mbox{Pr}( Y = 1 \mid X = x)
\]</span></p>
<p>As an example, let’s provide a prediction for a student that is 66 inches tall. What is the conditional probability of being female if you are 66 inches tall? In our dataset, we can estimate this by rounding to the nearest inch and computing:</p>
<div class="sourceCode" id="cb571"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb571-1" data-line-number="1">train_set <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb571-2" data-line-number="2"><span class="st">  </span><span class="kw">filter</span>(<span class="kw">round</span>(height)<span class="op">==</span><span class="dv">66</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb571-3" data-line-number="3"><span class="st">  </span><span class="kw">summarize</span>(<span class="dt">y_hat =</span> <span class="kw">mean</span>(sex<span class="op">==</span><span class="st">&quot;Female&quot;</span>))</a></code></pre></div>
<pre><code>  y_hat
1 0.327</code></pre>
<p>To construct a prediction algorithm, we want to estimate the proportion of the population that is female for any given height <span class="math inline">\(X=x\)</span>, which we write as the conditional probability described above: <span class="math inline">\(\mbox{Pr}( Y = 1 | X=x)\)</span>. Let’s see what this looks like for several values of <span class="math inline">\(x\)</span> (we will remove strata of <span class="math inline">\(x\)</span> with few data points):</p>
<div class="sourceCode" id="cb573"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb573-1" data-line-number="1">heights <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb573-2" data-line-number="2"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">x =</span> <span class="kw">round</span>(height)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb573-3" data-line-number="3"><span class="st">  </span><span class="kw">group_by</span>(x) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb573-4" data-line-number="4"><span class="st">  </span><span class="kw">filter</span>(<span class="kw">n</span>() <span class="op">&gt;=</span><span class="st"> </span><span class="dv">10</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb573-5" data-line-number="5"><span class="st">  </span><span class="kw">summarize</span>(<span class="dt">prop =</span> <span class="kw">mean</span>(sex <span class="op">==</span><span class="st"> &quot;Female&quot;</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb573-6" data-line-number="6"><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(x, prop)) <span class="op">+</span></a>
<a class="sourceLine" id="cb573-7" data-line-number="7"><span class="st">  </span><span class="kw">geom_point</span>()</a></code></pre></div>
<p><img src="Bayesian-Data-Science_files/figure-html/height-and-sex-conditional-probabilities-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>Since the results from the plot above look close to linear, and it is the only approach we currently know, we will try regression. We assume that:</p>
<p><span class="math display">\[p(x) = \mbox{Pr}( Y = 1 | X=x)  = \beta_0 + \beta_1 x\]</span></p>
<p>Note: because <span class="math inline">\(p_0(x) = 1 - p_1(x)\)</span>, we will only estimate <span class="math inline">\(p_1(x)\)</span> and drop the <span class="math inline">\(_1\)</span> index.</p>
<p>If we convert the factors to 0s and 1s, we can estimate <span class="math inline">\(\beta_0\)</span> and <span class="math inline">\(\beta_1\)</span> with least squares.</p>
<div class="sourceCode" id="cb574"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb574-1" data-line-number="1">lm_fit &lt;-<span class="st"> </span><span class="kw">mutate</span>(train_set, <span class="dt">y =</span> <span class="kw">as.numeric</span>(sex <span class="op">==</span><span class="st"> &quot;Female&quot;</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb574-2" data-line-number="2"><span class="st">  </span><span class="kw">lm</span>(y <span class="op">~</span><span class="st"> </span>height, <span class="dt">data =</span> .)</a></code></pre></div>
<p>Once we have estimates <span class="math inline">\(\hat{\beta}_0\)</span> and <span class="math inline">\(\hat{\beta}_1\)</span>, we can obtain an actual prediction. Our estimate of the conditional probability <span class="math inline">\(p(x)\)</span> is:</p>
<p><span class="math display">\[
\hat{p}(x) = \hat{\beta}_0+ \hat{\beta}_1 x
\]</span></p>
<p>To form a prediction, we define a <em>decision rule</em>: predict female if <span class="math inline">\(\hat{p}(x) &gt; 0.5\)</span>. We can compare our predictions to the outcomes using:</p>
<div class="sourceCode" id="cb575"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb575-1" data-line-number="1">p_hat &lt;-<span class="st"> </span><span class="kw">predict</span>(lm_fit, test_set)</a>
<a class="sourceLine" id="cb575-2" data-line-number="2">y_hat &lt;-<span class="st"> </span><span class="kw">ifelse</span>(p_hat <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>, <span class="st">&quot;Female&quot;</span>, <span class="st">&quot;Male&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">factor</span>()</a>
<a class="sourceLine" id="cb575-3" data-line-number="3"><span class="kw">confusionMatrix</span>(y_hat, test_set<span class="op">$</span>sex)[[<span class="st">&quot;Accuracy&quot;</span>]]</a></code></pre></div>
<pre><code>NULL</code></pre>
<p>We see this method does substantially better than guessing.</p>
<div id="generalized-linear-models" class="section level3">
<h3><span class="header-section-number">13.3.1</span> Generalized linear models</h3>
<p>The function <span class="math inline">\(\beta_0 + \beta_1 x\)</span> can take any value including negatives and values larger than 1. In fact, the estimate <span class="math inline">\(\hat{p}(x)\)</span> computed in the linear regression section does indeed become negative at around 76 inches.</p>
<div class="sourceCode" id="cb577"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb577-1" data-line-number="1">heights <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb577-2" data-line-number="2"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">x =</span> <span class="kw">round</span>(height)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb577-3" data-line-number="3"><span class="st">  </span><span class="kw">group_by</span>(x) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb577-4" data-line-number="4"><span class="st">  </span><span class="kw">filter</span>(<span class="kw">n</span>() <span class="op">&gt;=</span><span class="st"> </span><span class="dv">10</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb577-5" data-line-number="5"><span class="st">  </span><span class="kw">summarize</span>(<span class="dt">prop =</span> <span class="kw">mean</span>(sex <span class="op">==</span><span class="st"> &quot;Female&quot;</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb577-6" data-line-number="6"><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(x, prop)) <span class="op">+</span></a>
<a class="sourceLine" id="cb577-7" data-line-number="7"><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb577-8" data-line-number="8"><span class="st">  </span><span class="kw">geom_abline</span>(<span class="dt">intercept =</span> lm_fit<span class="op">$</span>coef[<span class="dv">1</span>], <span class="dt">slope =</span> lm_fit<span class="op">$</span>coef[<span class="dv">2</span>])</a></code></pre></div>
<p><img src="Bayesian-Data-Science_files/figure-html/regression-prediction-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>The range is:</p>
<div class="sourceCode" id="cb578"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb578-1" data-line-number="1"><span class="kw">range</span>(p_hat)</a></code></pre></div>
<pre><code>[1] -0.331  1.036</code></pre>
<p>But we are estimating a probability: <span class="math inline">\(\mbox{Pr}( Y = 1 \mid X = x)\)</span> which is constrained between 0 and 1.</p>
<p>The idea of generalized linear models (GLM) is to 1) define a distribution of <span class="math inline">\(Y\)</span> that is consistent with it’s possible outcomes and
2) find a function <span class="math inline">\(g\)</span> so that <span class="math inline">\(g(\mbox{Pr}( Y = 1 \mid X = x))\)</span> can be modeled as a linear combination of predictors.
Logistic regression is the most commonly used GLM. It is an extension of linear regression that assures that the estimate of <span class="math inline">\(\mbox{Pr}( Y = 1 \mid X = x)\)</span> is between 0 and 1. This approach makes use of the <em>logistic</em> transformation introduced in Section <a href="#logit"><strong>??</strong></a>:</p>
<p><span class="math display">\[ g(p) = \log \frac{p}{1-p}\]</span></p>
<p>This logistic transformation converts probability to log odds. As discussed in the data visualization lecture, the odds tell us how much more likely it is something will happen compared to not happening. <span class="math inline">\(p=0.5\)</span> means the odds are 1 to 1, thus the odds are 1. If <span class="math inline">\(p=0.75\)</span>, the odds are 3 to 1. A nice characteristic of this transformation is that it converts probabilities to be symmetric around 0. Here is a plot of <span class="math inline">\(g(p)\)</span> versus <span class="math inline">\(p\)</span>:</p>
<p><img src="Bayesian-Data-Science_files/figure-html/p-versus-logistic-of-p-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>With <em>logistic regression</em>, we model the conditional probability directly with:</p>
<p><span class="math display">\[ 
g\left\{ \mbox{Pr}(Y = 1 \mid X=x) \right\} = \beta_0 + \beta_1 x
\]</span></p>
<p>With this model, we can no longer use least squares. Instead we compute the <em>maximum likelihood estimate</em> (MLE). You can learn more about this concept in a statistical theory textbook<a href="#fn15" class="footnote-ref" id="fnref15"><sup>15</sup></a>.</p>
<p>In R, we can fit the logistic regression model with the function <code>glm</code>: generalized linear models. This function is more general than logistic regression so we need to specify the model we want through the <code>family</code> parameter:</p>
<div class="sourceCode" id="cb580"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb580-1" data-line-number="1">glm_fit &lt;-<span class="st"> </span>train_set <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb580-2" data-line-number="2"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">y =</span> <span class="kw">as.numeric</span>(sex <span class="op">==</span><span class="st"> &quot;Female&quot;</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb580-3" data-line-number="3"><span class="st">  </span><span class="kw">glm</span>(y <span class="op">~</span><span class="st"> </span>height, <span class="dt">data=</span>., <span class="dt">family =</span> <span class="st">&quot;binomial&quot;</span>)</a></code></pre></div>
<p>We can obtain prediction using the predict function:</p>
<div class="sourceCode" id="cb581"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb581-1" data-line-number="1">p_hat_logit &lt;-<span class="st"> </span><span class="kw">predict</span>(glm_fit, <span class="dt">newdata =</span> test_set, <span class="dt">type =</span> <span class="st">&quot;response&quot;</span>)</a></code></pre></div>
<p>When using <code>predict</code> with a <code>glm</code> object, we have to specify that we want <code>type=&quot;response&quot;</code> if we want the conditional probabilities, since the default is to return the logistic transformed values.</p>
<p>This model fits the data slightly better than the line:</p>
<p><img src="Bayesian-Data-Science_files/figure-html/conditional-prob-glm-fit-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>Because we have an estimate <span class="math inline">\(\hat{p}(x)\)</span>, we can obtain predictions:</p>
<div class="sourceCode" id="cb582"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb582-1" data-line-number="1">y_hat_logit &lt;-<span class="st"> </span><span class="kw">ifelse</span>(p_hat_logit <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>, <span class="st">&quot;Female&quot;</span>, <span class="st">&quot;Male&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span>factor</a>
<a class="sourceLine" id="cb582-2" data-line-number="2"><span class="kw">confusionMatrix</span>(y_hat_logit, test_set<span class="op">$</span>sex)[[<span class="st">&quot;Accuracy&quot;</span>]]</a></code></pre></div>
<pre><code>NULL</code></pre>
<p>The resulting predictions are similar. This is because the two estimates of <span class="math inline">\(p(x)\)</span> are larger than 1/2 in about the same region of x:</p>
<div class="sourceCode" id="cb584"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb584-1" data-line-number="1"><span class="kw">data.frame</span>(<span class="dt">x =</span> <span class="kw">seq</span>(<span class="kw">min</span>(tmp<span class="op">$</span>x), <span class="kw">max</span>(tmp<span class="op">$</span>x))) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb584-2" data-line-number="2"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">logistic =</span> <span class="kw">plogis</span>(glm_fit<span class="op">$</span>coef[<span class="dv">1</span>] <span class="op">+</span><span class="st"> </span>glm_fit<span class="op">$</span>coef[<span class="dv">2</span>]<span class="op">*</span>x),</a>
<a class="sourceLine" id="cb584-3" data-line-number="3">         <span class="dt">regression =</span> lm_fit<span class="op">$</span>coef[<span class="dv">1</span>] <span class="op">+</span><span class="st"> </span>lm_fit<span class="op">$</span>coef[<span class="dv">2</span>]<span class="op">*</span>x) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb584-4" data-line-number="4"><span class="st">  </span><span class="kw">gather</span>(method, p_x, <span class="op">-</span>x) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb584-5" data-line-number="5"><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(x, p_x, <span class="dt">color =</span> method)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb584-6" data-line-number="6"><span class="st">  </span><span class="kw">geom_line</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb584-7" data-line-number="7"><span class="st">  </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="fl">0.5</span>, <span class="dt">lty =</span> <span class="dv">5</span>)</a></code></pre></div>
<p><img src="Bayesian-Data-Science_files/figure-html/glm-prediction-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>Both linear and logistic regressions provide an estimate for the conditional expectation:</p>
<p><span class="math display">\[
\mbox{E}(Y \mid X=x)
\]</span>
which in the case of binary data is equivalent to the conditional probability:</p>
<p><span class="math display">\[
\mbox{Pr}(Y = 1 \mid X = x)
\]</span></p>
</div>
<div id="logistic-regression-with-more-than-one-predictor" class="section level3">
<h3><span class="header-section-number">13.3.2</span> Logistic regression with more than one predictor</h3>
<p>In this section we apply logistic regression to the two or seven data introduced in Section <a href="11-introduction-to-machine-learning.html#two-or-seven">11.8</a>. In this case, we are interested in estimating a conditional probability that depends on two variables. The standard logistic regression model in this case will assume that</p>
<p><span class="math display">\[
g\{p(x_1, x_2)\}= g\{\mbox{Pr}(Y=1 \mid X_1=x_1 , X_2 = x_2)\} = 
\beta_0 + \beta_1 x_1 + \beta_2 x_2
\]</span>
with <span class="math inline">\(g(p) = \log \frac{p}{1-p}\)</span> the logistic function described in the previous section. To fit the model we use the following code:</p>
<div class="sourceCode" id="cb585"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb585-1" data-line-number="1">fit_glm &lt;-<span class="st"> </span><span class="kw">glm</span>(y <span class="op">~</span><span class="st"> </span>x_<span class="dv">1</span> <span class="op">+</span><span class="st"> </span>x_<span class="dv">2</span>, <span class="dt">data=</span>mnist_<span class="dv">27</span><span class="op">$</span>train, <span class="dt">family =</span> <span class="st">&quot;binomial&quot;</span>)</a>
<a class="sourceLine" id="cb585-2" data-line-number="2">p_hat_glm &lt;-<span class="st"> </span><span class="kw">predict</span>(fit_glm, mnist_<span class="dv">27</span><span class="op">$</span>test)</a>
<a class="sourceLine" id="cb585-3" data-line-number="3">y_hat_glm &lt;-<span class="st"> </span><span class="kw">factor</span>(<span class="kw">ifelse</span>(p_hat_glm <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>, <span class="dv">7</span>, <span class="dv">2</span>))</a>
<a class="sourceLine" id="cb585-4" data-line-number="4"><span class="kw">confusionMatrix</span>(y_hat_glm, mnist_<span class="dv">27</span><span class="op">$</span>test<span class="op">$</span>y)<span class="op">$</span>overall[<span class="st">&quot;Accuracy&quot;</span>]</a></code></pre></div>
<pre><code>Accuracy 
    0.76 </code></pre>
<p>Comparing to the results we obtained in Section <a href="11-introduction-to-machine-learning.html#two-or-seven">11.8</a>, we see that logistic regression performs similarly to regression.
This is not surprising, given that the estimate of <span class="math inline">\(\hat{p}(x_1, x_2)\)</span> looks similar as well:</p>
<div class="sourceCode" id="cb587"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb587-1" data-line-number="1">p_hat &lt;-<span class="st"> </span><span class="kw">predict</span>(fit_glm, <span class="dt">newdata =</span> mnist_<span class="dv">27</span><span class="op">$</span>true_p, <span class="dt">type =</span> <span class="st">&quot;response&quot;</span>)</a>
<a class="sourceLine" id="cb587-2" data-line-number="2">mnist_<span class="dv">27</span><span class="op">$</span>true_p <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">mutate</span>(<span class="dt">p_hat =</span> p_hat) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb587-3" data-line-number="3"><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(x_<span class="dv">1</span>, x_<span class="dv">2</span>,  <span class="dt">z=</span>p_hat, <span class="dt">fill=</span>p_hat)) <span class="op">+</span></a>
<a class="sourceLine" id="cb587-4" data-line-number="4"><span class="st">  </span><span class="kw">geom_raster</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb587-5" data-line-number="5"><span class="st">  </span><span class="kw">scale_fill_gradientn</span>(<span class="dt">colors=</span><span class="kw">c</span>(<span class="st">&quot;#F8766D&quot;</span>,<span class="st">&quot;white&quot;</span>,<span class="st">&quot;#00BFC4&quot;</span>)) <span class="op">+</span></a>
<a class="sourceLine" id="cb587-6" data-line-number="6"><span class="st">  </span><span class="kw">stat_contour</span>(<span class="dt">breaks=</span><span class="kw">c</span>(<span class="fl">0.5</span>), <span class="dt">color=</span><span class="st">&quot;black&quot;</span>) </a></code></pre></div>
<p><img src="Bayesian-Data-Science_files/figure-html/logistic-p-hat-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>Just like regression, the decision rule is a line, a fact that can be corroborated mathematically since</p>
<p><span class="math display">\[
g^{-1}(\hat{\beta}_0 + \hat{\beta}_1 x_1 + \hat{\beta}_2 x_2) = 0.5 \implies
\hat{\beta}_0 + \hat{\beta}_1 x_1 + \hat{\beta}_2 x_2 = g(0.5) = 0 \implies
x_2 = -\hat{\beta}_0/\hat{\beta}_2 -\hat{\beta}_1/\hat{\beta}_2 x_1
\]</span></p>
<p>Thus <span class="math inline">\(x_2\)</span> is a linear function of <span class="math inline">\(x_1\)</span>. This implies that, just like regression, our logistic regression approach has no chance of capturing the non-linear nature of the true <span class="math inline">\(p(x_1,x_2)\)</span>. Once we move on to more complex examples, we will see that linear regression and generalized linear regression are limited and not flexible enough to be useful for most machine learning challenges. The new techniques we learn are essentially approaches to estimating the conditional probability in a way that is more flexible.</p>
</div>
</div>
<div id="exercises-6" class="section level2">
<h2><span class="header-section-number">13.4</span> Exercises</h2>
<p>1. Define the following dataset:</p>
<div class="sourceCode" id="cb588"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb588-1" data-line-number="1">make_data &lt;-<span class="st"> </span><span class="cf">function</span>(<span class="dt">n =</span> <span class="dv">1000</span>, <span class="dt">p =</span> <span class="fl">0.5</span>, </a>
<a class="sourceLine" id="cb588-2" data-line-number="2">                      <span class="dt">mu_0 =</span> <span class="dv">0</span>, <span class="dt">mu_1 =</span> <span class="dv">2</span>, </a>
<a class="sourceLine" id="cb588-3" data-line-number="3">                      <span class="dt">sigma_0 =</span> <span class="dv">1</span>,  <span class="dt">sigma_1 =</span> <span class="dv">1</span>){</a>
<a class="sourceLine" id="cb588-4" data-line-number="4">  y &lt;-<span class="st"> </span><span class="kw">rbinom</span>(n, <span class="dv">1</span>, p)</a>
<a class="sourceLine" id="cb588-5" data-line-number="5">  f_<span class="dv">0</span> &lt;-<span class="st"> </span><span class="kw">rnorm</span>(n, mu_<span class="dv">0</span>, sigma_<span class="dv">0</span>)</a>
<a class="sourceLine" id="cb588-6" data-line-number="6">  f_<span class="dv">1</span> &lt;-<span class="st"> </span><span class="kw">rnorm</span>(n, mu_<span class="dv">1</span>, sigma_<span class="dv">1</span>)</a>
<a class="sourceLine" id="cb588-7" data-line-number="7">  x &lt;-<span class="st"> </span><span class="kw">ifelse</span>(y <span class="op">==</span><span class="st"> </span><span class="dv">1</span>, f_<span class="dv">1</span>, f_<span class="dv">0</span>)</a>
<a class="sourceLine" id="cb588-8" data-line-number="8">  test_index &lt;-<span class="st"> </span><span class="kw">createDataPartition</span>(y, <span class="dt">times =</span> <span class="dv">1</span>, <span class="dt">p =</span> <span class="fl">0.5</span>, <span class="dt">list =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb588-9" data-line-number="9">  <span class="kw">list</span>(<span class="dt">train =</span> <span class="kw">data.frame</span>(<span class="dt">x =</span> x, <span class="dt">y =</span> <span class="kw">as.factor</span>(y)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb588-10" data-line-number="10"><span class="st">         </span><span class="kw">slice</span>(<span class="op">-</span>test_index),</a>
<a class="sourceLine" id="cb588-11" data-line-number="11">       <span class="dt">test =</span> <span class="kw">data.frame</span>(<span class="dt">x =</span> x, <span class="dt">y =</span> <span class="kw">as.factor</span>(y)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb588-12" data-line-number="12"><span class="st">         </span><span class="kw">slice</span>(test_index))</a>
<a class="sourceLine" id="cb588-13" data-line-number="13">}</a>
<a class="sourceLine" id="cb588-14" data-line-number="14">dat &lt;-<span class="st"> </span><span class="kw">make_data</span>()</a></code></pre></div>
<p>Note that we have defined a variable <code>x</code> that is predictive of a binary outcome <code>y</code>.</p>
<div class="sourceCode" id="cb589"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb589-1" data-line-number="1">dat<span class="op">$</span>train <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(x, <span class="dt">color =</span> y)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_density</span>()</a></code></pre></div>
<p>Compare the accuracy of linear regression and logistic regression.</p>
<p>2. Repeat the simulation from exercise 1 100 times and compare the average accuracy for each method and notice they give practically the same answer.</p>
<p>3. Generate 25 different datasets changing the difference between the two class: <code>delta &lt;- seq(0, 3, len = 25)</code>. Plot accuracy versus <code>delta</code>.</p>

</div>
<div id="k-nearest-neighbors" class="section level2">
<h2><span class="header-section-number">13.5</span> k-nearest neighbors</h2>
<p>We introduced the kNN algorithm in Section <a href="12-cross-validation.html#knn-cv-intro">12.1</a> and demonstrated how we use cross validation to pick <span class="math inline">\(k\)</span> in Section <a href="#caret-cv"><strong>??</strong></a>. Here we quickly review how we fit a kNN model using the <strong>caret</strong> package. In Section <a href="#caret-cv"><strong>??</strong></a> we introduced the following code to fit a kNN model:</p>
<div class="sourceCode" id="cb590"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb590-1" data-line-number="1">train_knn &lt;-<span class="st"> </span><span class="kw">train</span>(y <span class="op">~</span><span class="st"> </span>., <span class="dt">method =</span> <span class="st">&quot;knn&quot;</span>, </a>
<a class="sourceLine" id="cb590-2" data-line-number="2">                   <span class="dt">data =</span> mnist_<span class="dv">27</span><span class="op">$</span>train,</a>
<a class="sourceLine" id="cb590-3" data-line-number="3">                   <span class="dt">tuneGrid =</span> <span class="kw">data.frame</span>(<span class="dt">k =</span> <span class="kw">seq</span>(<span class="dv">9</span>, <span class="dv">71</span>, <span class="dv">2</span>)))</a></code></pre></div>
<p>We saw that the parameter that maximized the estimated accuracy was:</p>
<div class="sourceCode" id="cb591"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb591-1" data-line-number="1">train_knn<span class="op">$</span>bestTune</a></code></pre></div>
<pre><code>    k
10 27</code></pre>
<p>This model improves the accuracy over regression and logistic regression:</p>
<div class="sourceCode" id="cb593"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb593-1" data-line-number="1"><span class="kw">confusionMatrix</span>(<span class="kw">predict</span>(train_knn, mnist_<span class="dv">27</span><span class="op">$</span>test, <span class="dt">type =</span> <span class="st">&quot;raw&quot;</span>),</a>
<a class="sourceLine" id="cb593-2" data-line-number="2">                mnist_<span class="dv">27</span><span class="op">$</span>test<span class="op">$</span>y)<span class="op">$</span>overall[<span class="st">&quot;Accuracy&quot;</span>]</a></code></pre></div>
<pre><code>Accuracy 
   0.835 </code></pre>
<p>A plot of the estimated conditional probability shows that the kNN estimate is flexible enough and does indeed capture the shape of the true conditional probability.</p>
<p><img src="Bayesian-Data-Science_files/figure-html/best-knn-fit-1.png" width="100%" style="display: block; margin: auto;" /></p>
</div>
<div id="exercises-7" class="section level2">
<h2><span class="header-section-number">13.6</span> Exercises</h2>
<p>1. Earlier we used logistic regression to predict sex from height. Use kNN to do the same. Use the code described in this chapter to select the <span class="math inline">\(F_1\)</span> measure and plot it against <span class="math inline">\(k\)</span>. Compare to the <span class="math inline">\(F_1\)</span> of about 0.6 we obtained with regression.</p>
<p>2. Load the following dataset:</p>
<div class="sourceCode" id="cb595"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb595-1" data-line-number="1"><span class="kw">data</span>(<span class="st">&quot;tissue_gene_expression&quot;</span>)</a></code></pre></div>
<p>This dataset includes a matrix <code>x</code>:</p>
<div class="sourceCode" id="cb596"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb596-1" data-line-number="1"><span class="kw">dim</span>(tissue_gene_expression<span class="op">$</span>x)</a></code></pre></div>
<p>with the gene expression measured on 500 genes for 189 biological samples representing seven different tissues. The tissue type is stored in <code>y</code>:</p>
<div class="sourceCode" id="cb597"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb597-1" data-line-number="1"><span class="kw">table</span>(tissue_gene_expression<span class="op">$</span>y)</a></code></pre></div>
<p>Split the data in training and test sets, then use kNN to predict tissue type and see what accuracy you obtain. Try it for <span class="math inline">\(k = 1, 3, \dots, 11\)</span>.</p>

</div>
<div id="generative-models" class="section level2">
<h2><span class="header-section-number">13.7</span> Generative models</h2>
<p>We have described how, when using squared loss, the conditional expectation/probabilities provide the best approach to developing a decision rule. In a binary case, the smallest true error we can achieve is determined by Bayes’ rule, which is a decision rule based on the true conditional probability:</p>
<p><span class="math display">\[
p(\mathbf{x}) = \mbox{Pr}(Y=1 \mid \mathbf{X}=\mathbf{x}) 
\]</span></p>
<p>We have described several approaches to estimating <span class="math inline">\(p(\mathbf{x})\)</span>. In all these approaches, we estimate the conditional probability directly and do not consider the distribution of the predictors. In machine learning, these are referred to as <em>discriminative</em> approaches.</p>
<p>However, Bayes’ theorem tells us that knowing the distribution of the predictors <span class="math inline">\(\mathbf{X}\)</span> may be useful. Methods that model the joint distribution of <span class="math inline">\(Y\)</span> and <span class="math inline">\(\mathbf{X}\)</span> are referred to as <em>generative models</em> (we model how the entire data, <span class="math inline">\(\mathbf{X}\)</span> and <span class="math inline">\(Y\)</span>, are generated). We start by describing the most general generative model, Naive Bayes, and then proceed to describe two specific cases, quadratic discriminant analysis (QDA) and linear discriminant analysis (LDA).</p>
<div id="naive-bayes" class="section level3">
<h3><span class="header-section-number">13.7.1</span> Naive Bayes</h3>
<p>Recall that Bayes rule tells us that we can rewrite <span class="math inline">\(p(\mathbf{x})\)</span> like this:</p>
<p><span class="math display">\[
p(\mathbf{x}) = \mbox{Pr}(Y=1|\mathbf{X}=\mathbf{x}) = \frac{f_{\mathbf{X}|Y=1}(\mathbf{x}) \mbox{Pr}(Y=1)}
{ f_{\mathbf{X}|Y=0}(\mathbf{x})\mbox{Pr}(Y=0)  + f_{\mathbf{X}|Y=1}(\mathbf{x})\mbox{Pr}(Y=1) }
\]</span></p>
<p>with <span class="math inline">\(f_{\mathbf{X}|Y=1}\)</span> and <span class="math inline">\(f_{\mathbf{X}|Y=0}\)</span> representing the distribution functions of the predictor <span class="math inline">\(\mathbf{X}\)</span> for the two classes <span class="math inline">\(Y=1\)</span> and <span class="math inline">\(Y=0\)</span>. The formula implies that if we can estimate these conditional distributions of the predictors, we can develop a powerful decision rule. However, this is a big <em>if</em>. As we go forward, we will encounter examples in which <span class="math inline">\(\mathbf{X}\)</span> has many dimensions and we do not have much information about the distribution. In these cases, Naive Bayes will be practically impossible to implement. However, there are instances in which we have a small number of predictors (not much more than 2) and many categories in which generative models can be quite powerful. We describe two specific examples and use our previously described case studies to illustrate them.</p>
<p>Let’s start with a very simple and uninteresting, yet illustrative, case: the example related to predicting sex from height.</p>
<div class="sourceCode" id="cb598"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb598-1" data-line-number="1"><span class="kw">library</span>(tidyverse)</a>
<a class="sourceLine" id="cb598-2" data-line-number="2"><span class="kw">library</span>(caret)</a>
<a class="sourceLine" id="cb598-3" data-line-number="3"></a>
<a class="sourceLine" id="cb598-4" data-line-number="4"><span class="kw">library</span>(dslabs)</a>
<a class="sourceLine" id="cb598-5" data-line-number="5"><span class="kw">data</span>(<span class="st">&quot;heights&quot;</span>)</a>
<a class="sourceLine" id="cb598-6" data-line-number="6"></a>
<a class="sourceLine" id="cb598-7" data-line-number="7">y &lt;-<span class="st"> </span>heights<span class="op">$</span>height</a>
<a class="sourceLine" id="cb598-8" data-line-number="8"><span class="kw">set.seed</span>(<span class="dv">1995</span>)</a>
<a class="sourceLine" id="cb598-9" data-line-number="9">test_index &lt;-<span class="st"> </span><span class="kw">createDataPartition</span>(y, <span class="dt">times =</span> <span class="dv">1</span>, <span class="dt">p =</span> <span class="fl">0.5</span>, <span class="dt">list =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb598-10" data-line-number="10">train_set &lt;-<span class="st"> </span>heights <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">slice</span>(<span class="op">-</span>test_index)</a>
<a class="sourceLine" id="cb598-11" data-line-number="11">test_set &lt;-<span class="st"> </span>heights <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">slice</span>(test_index)</a></code></pre></div>
<p>In this case, the Naive Bayes approach is particularly appropriate because we know that the normal distribution is a good approximation for the conditional distributions of height given sex for both classes <span class="math inline">\(Y=1\)</span> (female) and <span class="math inline">\(Y=0\)</span> (male). This implies that we can approximate the conditional distributions <span class="math inline">\(f_{X|Y=1}\)</span> and <span class="math inline">\(f_{X|Y=0}\)</span> by simply estimating averages and standard deviations from the data:</p>
<div class="sourceCode" id="cb599"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb599-1" data-line-number="1">params &lt;-<span class="st"> </span>train_set <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb599-2" data-line-number="2"><span class="st">  </span><span class="kw">group_by</span>(sex) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb599-3" data-line-number="3"><span class="st">  </span><span class="kw">summarize</span>(<span class="dt">avg =</span> <span class="kw">mean</span>(height), <span class="dt">sd =</span> <span class="kw">sd</span>(height))</a>
<a class="sourceLine" id="cb599-4" data-line-number="4">params</a></code></pre></div>
<pre><code># A tibble: 2 x 3
  sex        avg      sd
  &lt;fct&gt;    &lt;dbl&gt;   &lt;dbl&gt;
1 Female 64.7647 4.13804
2 Male   69.2269 3.56905</code></pre>
<p>The prevalence, which we will denote with <span class="math inline">\(\pi = \mbox{Pr}(Y=1)\)</span>, can be estimated from the data with:</p>
<div class="sourceCode" id="cb601"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb601-1" data-line-number="1">pi &lt;-<span class="st"> </span>train_set <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">summarize</span>(<span class="dt">pi=</span><span class="kw">mean</span>(sex<span class="op">==</span><span class="st">&quot;Female&quot;</span>)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">pull</span>(pi)</a>
<a class="sourceLine" id="cb601-2" data-line-number="2">pi</a></code></pre></div>
<pre><code>[1] 0.212</code></pre>
<p>Now we can use our estimates of average and standard deviation to get an actual rule:</p>
<div class="sourceCode" id="cb603"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb603-1" data-line-number="1">x &lt;-<span class="st"> </span>test_set<span class="op">$</span>height</a>
<a class="sourceLine" id="cb603-2" data-line-number="2"></a>
<a class="sourceLine" id="cb603-3" data-line-number="3">f0 &lt;-<span class="st"> </span><span class="kw">dnorm</span>(x, params<span class="op">$</span>avg[<span class="dv">2</span>], params<span class="op">$</span>sd[<span class="dv">2</span>])</a>
<a class="sourceLine" id="cb603-4" data-line-number="4">f1 &lt;-<span class="st"> </span><span class="kw">dnorm</span>(x, params<span class="op">$</span>avg[<span class="dv">1</span>], params<span class="op">$</span>sd[<span class="dv">1</span>])</a>
<a class="sourceLine" id="cb603-5" data-line-number="5"></a>
<a class="sourceLine" id="cb603-6" data-line-number="6">p_hat_bayes &lt;-<span class="st"> </span>f1<span class="op">*</span>pi <span class="op">/</span><span class="st"> </span>(f1<span class="op">*</span>pi <span class="op">+</span><span class="st"> </span>f0<span class="op">*</span>(<span class="dv">1</span> <span class="op">-</span><span class="st"> </span>pi))</a></code></pre></div>
<p>Our Naive Bayes estimate <span class="math inline">\(\hat{p}(x)\)</span> looks a lot like our logistic regression estimate:</p>
<p><img src="Bayesian-Data-Science_files/figure-html/conditional-prob-glm-fit-2-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>In fact, we can show that the Naive Bayes approach is similar to the logistic regression prediction mathematically. However, we leave the demonstration to a more advanced text, such as the Elements of Statistical Learning<a href="#fn16" class="footnote-ref" id="fnref16"><sup>16</sup></a>. We can see that they are similar empirically by comparing the two resulting curves.</p>
</div>
<div id="controlling-prevalence" class="section level3">
<h3><span class="header-section-number">13.7.2</span> Controlling prevalence</h3>
<p>One useful feature of the Naive Bayes approach is that it includes a parameter to account for differences in prevalence. Using our sample, we estimated <span class="math inline">\(f_{X|Y=1}\)</span>, <span class="math inline">\(f_{X|Y=0}\)</span> and <span class="math inline">\(\pi\)</span>. If we use hats to denote the estimates, we can write <span class="math inline">\(\hat{p}(x)\)</span> as:</p>
<p><span class="math display">\[
\hat{p}(x)= \frac{\hat{f}_{X|Y=1}(x) \hat{\pi}}
{ \hat{f}_{X|Y=0}(x)(1-\hat{\pi}) + \hat{f}_{X|Y=1}(x)\hat{\pi} }
\]</span></p>
<p>As we discussed earlier, our sample has a much lower prevalence, 0.21, than the general population. So if we use the rule <span class="math inline">\(\hat{p}(x)&gt;0.5\)</span> to predict females, our accuracy will be affected due to the low sensitivity:</p>
<div class="sourceCode" id="cb604"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb604-1" data-line-number="1">y_hat_bayes &lt;-<span class="st"> </span><span class="kw">ifelse</span>(p_hat_bayes <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>, <span class="st">&quot;Female&quot;</span>, <span class="st">&quot;Male&quot;</span>)</a>
<a class="sourceLine" id="cb604-2" data-line-number="2"><span class="kw">sensitivity</span>(<span class="dt">data =</span> <span class="kw">factor</span>(y_hat_bayes), <span class="dt">reference =</span> <span class="kw">factor</span>(test_set<span class="op">$</span>sex))</a></code></pre></div>
<pre><code>[1] 0.213</code></pre>
<p>Again, this is because the algorithm gives more weight to specificity to account for the low prevalence:</p>
<div class="sourceCode" id="cb606"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb606-1" data-line-number="1"><span class="kw">specificity</span>(<span class="dt">data =</span> <span class="kw">factor</span>(y_hat_bayes), <span class="dt">reference =</span> <span class="kw">factor</span>(test_set<span class="op">$</span>sex))</a></code></pre></div>
<pre><code>[1] 0.967</code></pre>
<p>This is due mainly to the fact that <span class="math inline">\(\hat{\pi}\)</span> is substantially less than 0.5, so we tend to predict <code>Male</code> more often. It makes sense for a machine learning algorithm to do this in our sample because we do have a higher percentage of males. But if we were to extrapolate this to a general population, our overall accuracy would be affected by the low sensitivity.</p>
<p>The Naive Bayes approach gives us a direct way to correct this since we can simply force <span class="math inline">\(\hat{\pi}\)</span> to be whatever value we want it to be. So to balance specificity and sensitivity, instead of changing the cutoff in the decision rule, we could simply change <span class="math inline">\(\hat{\pi}\)</span> to 0.5 like this:</p>
<div class="sourceCode" id="cb608"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb608-1" data-line-number="1">p_hat_bayes_unbiased &lt;-<span class="st"> </span>f1 <span class="op">*</span><span class="st"> </span><span class="fl">0.5</span> <span class="op">/</span><span class="st"> </span>(f1 <span class="op">*</span><span class="st"> </span><span class="fl">0.5</span> <span class="op">+</span><span class="st"> </span>f0 <span class="op">*</span><span class="st"> </span>(<span class="dv">1</span> <span class="op">-</span><span class="st"> </span><span class="fl">0.5</span>)) </a>
<a class="sourceLine" id="cb608-2" data-line-number="2">y_hat_bayes_unbiased &lt;-<span class="st"> </span><span class="kw">ifelse</span>(p_hat_bayes_unbiased<span class="op">&gt;</span><span class="st"> </span><span class="fl">0.5</span>, <span class="st">&quot;Female&quot;</span>, <span class="st">&quot;Male&quot;</span>)</a></code></pre></div>
<p>Note the difference in sensitivity with a better balance:</p>
<div class="sourceCode" id="cb609"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb609-1" data-line-number="1"><span class="kw">sensitivity</span>(<span class="kw">factor</span>(y_hat_bayes_unbiased), <span class="kw">factor</span>(test_set<span class="op">$</span>sex))</a></code></pre></div>
<pre><code>[1] 0.693</code></pre>
<div class="sourceCode" id="cb611"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb611-1" data-line-number="1"><span class="kw">specificity</span>(<span class="kw">factor</span>(y_hat_bayes_unbiased), <span class="kw">factor</span>(test_set<span class="op">$</span>sex))</a></code></pre></div>
<pre><code>[1] 0.832</code></pre>
<p>The new rule also gives us a very intuitive cutoff between 66-67, which is about the middle of the female and male average heights:</p>
<div class="sourceCode" id="cb613"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb613-1" data-line-number="1"><span class="kw">qplot</span>(x, p_hat_bayes_unbiased, <span class="dt">geom =</span> <span class="st">&quot;line&quot;</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb613-2" data-line-number="2"><span class="st">  </span><span class="kw">geom_hline</span>(<span class="dt">yintercept =</span> <span class="fl">0.5</span>, <span class="dt">lty =</span> <span class="dv">2</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb613-3" data-line-number="3"><span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> <span class="dv">67</span>, <span class="dt">lty =</span> <span class="dv">2</span>)</a></code></pre></div>
<p><img src="Bayesian-Data-Science_files/figure-html/naive-with-good-prevalence-1.png" width="70%" style="display: block; margin: auto;" /></p>
</div>
<div id="quadratic-discriminant-analysis" class="section level3">
<h3><span class="header-section-number">13.7.3</span> Quadratic discriminant analysis</h3>
<p>Quadratic Discriminant Analysis (QDA) is a version of Naive Bayes in which we assume that the distributions <span class="math inline">\(p_{\mathbf{X}|Y=1}(x)\)</span> and <span class="math inline">\(p_{\mathbf{X}|Y=0}(\mathbf{x})\)</span> are multivariate normal. The simple example we described in the previous section is actually QDA. Let’s now look at a slightly more complicated case: the 2 or 7 example.</p>
<div class="sourceCode" id="cb614"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb614-1" data-line-number="1"><span class="kw">data</span>(<span class="st">&quot;mnist_27&quot;</span>)</a></code></pre></div>
<p>In this case, we have two predictors so we assume each one is bivariate normal. This implies that we need to estimate two averages, two standard deviations, and a correlation for each case <span class="math inline">\(Y=1\)</span> and <span class="math inline">\(Y=0\)</span>. Once we have these, we can approximate the distributions <span class="math inline">\(f_{X_1,X_2|Y=1}\)</span> and <span class="math inline">\(f_{X_1, X_2|Y=0}\)</span>. We can easily estimate parameters from the data:</p>
<div class="sourceCode" id="cb615"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb615-1" data-line-number="1">params &lt;-<span class="st"> </span>mnist_<span class="dv">27</span><span class="op">$</span>train <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb615-2" data-line-number="2"><span class="st">  </span><span class="kw">group_by</span>(y) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb615-3" data-line-number="3"><span class="st">  </span><span class="kw">summarize</span>(<span class="dt">avg_1 =</span> <span class="kw">mean</span>(x_<span class="dv">1</span>), <span class="dt">avg_2 =</span> <span class="kw">mean</span>(x_<span class="dv">2</span>), </a>
<a class="sourceLine" id="cb615-4" data-line-number="4">            <span class="dt">sd_1=</span> <span class="kw">sd</span>(x_<span class="dv">1</span>), <span class="dt">sd_2 =</span> <span class="kw">sd</span>(x_<span class="dv">2</span>), </a>
<a class="sourceLine" id="cb615-5" data-line-number="5">            <span class="dt">r =</span> <span class="kw">cor</span>(x_<span class="dv">1</span>, x_<span class="dv">2</span>))</a>
<a class="sourceLine" id="cb615-6" data-line-number="6">params</a></code></pre></div>
<pre><code># A tibble: 2 x 6
  y        avg_1    avg_2      sd_1      sd_2        r
  &lt;fct&gt;    &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 2     0.128780 0.283154 0.0701748 0.0577989 0.400750
2 7     0.234113 0.288186 0.0718801 0.104796  0.455161</code></pre>
<p>Here we provide a visual way of showing the approach. We plot the data and use contour plots to give an idea of what the two estimated normal densities look like (we show the curve representing a region that includes 95% of the points):</p>
<div class="sourceCode" id="cb617"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb617-1" data-line-number="1">mnist_<span class="dv">27</span><span class="op">$</span>train <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">mutate</span>(<span class="dt">y =</span> <span class="kw">factor</span>(y)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb617-2" data-line-number="2"><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(x_<span class="dv">1</span>, x_<span class="dv">2</span>, <span class="dt">fill =</span> y, <span class="dt">color=</span>y)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb617-3" data-line-number="3"><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">show.legend =</span> <span class="ot">FALSE</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb617-4" data-line-number="4"><span class="st">  </span><span class="kw">stat_ellipse</span>(<span class="dt">type=</span><span class="st">&quot;norm&quot;</span>, <span class="dt">lwd =</span> <span class="fl">1.5</span>)</a></code></pre></div>
<p><img src="Bayesian-Data-Science_files/figure-html/qda-explained-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>This defines the following estimate of <span class="math inline">\(f(x_1, x_2)\)</span>.</p>
<p>We can use the <code>train</code> function from the <strong>caret</strong> package to fit the model and obtain predictors:</p>
<div class="sourceCode" id="cb618"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb618-1" data-line-number="1"><span class="kw">library</span>(caret)</a>
<a class="sourceLine" id="cb618-2" data-line-number="2">train_qda &lt;-<span class="st"> </span><span class="kw">train</span>(y <span class="op">~</span><span class="st"> </span>., <span class="dt">method =</span> <span class="st">&quot;qda&quot;</span>, <span class="dt">data =</span> mnist_<span class="dv">27</span><span class="op">$</span>train)</a></code></pre></div>
<p>We see that we obtain relatively good accuracy:</p>
<div class="sourceCode" id="cb619"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb619-1" data-line-number="1">y_hat &lt;-<span class="st"> </span><span class="kw">predict</span>(train_qda, mnist_<span class="dv">27</span><span class="op">$</span>test)</a>
<a class="sourceLine" id="cb619-2" data-line-number="2"><span class="kw">confusionMatrix</span>(y_hat, mnist_<span class="dv">27</span><span class="op">$</span>test<span class="op">$</span>y)<span class="op">$</span>overall[<span class="st">&quot;Accuracy&quot;</span>]</a></code></pre></div>
<pre><code>Accuracy 
    0.82 </code></pre>
<p>The estimated conditional probability looks relatively good, although it does not fit as well as the kernel smoothers:</p>
<p><img src="Bayesian-Data-Science_files/figure-html/qda-estimate-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>One reason QDA does not work as well as the kernel methods is perhaps because the assumption of normality does not quite hold. Although for the 2s it seems reasonable, for the 7s it does seem to be off. Notice the slight curvature in the points for the 7s:</p>
<div class="sourceCode" id="cb621"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb621-1" data-line-number="1">mnist_<span class="dv">27</span><span class="op">$</span>train <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">mutate</span>(<span class="dt">y =</span> <span class="kw">factor</span>(y)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb621-2" data-line-number="2"><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(x_<span class="dv">1</span>, x_<span class="dv">2</span>, <span class="dt">fill =</span> y, <span class="dt">color=</span>y)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb621-3" data-line-number="3"><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">show.legend =</span> <span class="ot">FALSE</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb621-4" data-line-number="4"><span class="st">  </span><span class="kw">stat_ellipse</span>(<span class="dt">type=</span><span class="st">&quot;norm&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb621-5" data-line-number="5"><span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span>y)</a></code></pre></div>
<p><img src="Bayesian-Data-Science_files/figure-html/qda-does-not-fit-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>QDA can work well here, but it becomes harder to use as the number of predictors increases. Here we have 2 predictors and had to compute 4 means, 4 SDs, and 2 correlations. How many parameters would we have if instead of 2 predictors, we had 10?
The main problem comes from estimating correlations for 10 predictors. With 10, we have 45 correlations for each class. In general, the formula is <span class="math inline">\(K\times p(p-1)/2\)</span>, which gets big fast. Once the number of parameters approaches the size of our data, the method becomes impractical due to overfitting.</p>
</div>
<div id="linear-discriminant-analysis" class="section level3">
<h3><span class="header-section-number">13.7.4</span> Linear discriminant analysis</h3>
<p>A relatively simple solution to the problem of having too many parameters is to assume that the correlation structure is the same for all classes, which reduces the number of parameters we need to estimate.</p>
<p>In this case, we would compute just one pair of standard deviations and one correlation,
<!--so the parameters would look something like this:


```r
params <- mnist_27$train %>% 
  group_by(y) %>% 
  summarize(avg_1 = mean(x_1), avg_2 = mean(x_2), 
            sd_1= sd(x_1), sd_2 = sd(x_2), 
            r = cor(x_1,x_2))

params <- params %>% mutate(sd_1 = mean(sd_1), sd_2=mean(sd_2), r=mean(r))
params 
```

```
# A tibble: 2 x 6
  y        avg_1    avg_2      sd_1      sd_2        r
  <fct>    <dbl>    <dbl>     <dbl>     <dbl>    <dbl>
1 2     0.128780 0.283154 0.0710274 0.0812973 0.427955
2 7     0.234113 0.288186 0.0710274 0.0812973 0.427955
```
-->
and the distributions looks like this:</p>
<p><img src="Bayesian-Data-Science_files/figure-html/lda-explained-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>Now the size of the ellipses as well as the angle are the same. This is because they have the same standard deviations and correlations.</p>
<p>We can fit the LDA model using <strong>caret</strong>:</p>
<div class="sourceCode" id="cb622"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb622-1" data-line-number="1">train_lda &lt;-<span class="st"> </span><span class="kw">train</span>(y <span class="op">~</span><span class="st"> </span>., <span class="dt">method =</span> <span class="st">&quot;lda&quot;</span>, <span class="dt">data =</span> mnist_<span class="dv">27</span><span class="op">$</span>train)</a>
<a class="sourceLine" id="cb622-2" data-line-number="2">y_hat &lt;-<span class="st"> </span><span class="kw">predict</span>(train_lda, mnist_<span class="dv">27</span><span class="op">$</span>test)</a>
<a class="sourceLine" id="cb622-3" data-line-number="3"><span class="kw">confusionMatrix</span>(y_hat, mnist_<span class="dv">27</span><span class="op">$</span>test<span class="op">$</span>y)<span class="op">$</span>overall[<span class="st">&quot;Accuracy&quot;</span>]</a></code></pre></div>
<pre><code>Accuracy 
    0.75 </code></pre>
<p>When we force this assumption, we can show mathematically that the boundary is a line, just as with logistic regression. For this reason, we call the method <em>linear</em> discriminant analysis (LDA). Similarly, for QDA, we can show that the boundary must be a quadratic function.</p>
<p><img src="Bayesian-Data-Science_files/figure-html/lda-estimate-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>In the case of LDA, the lack of flexibility does not permit us to capture the non-linearity in the true conditional probability function.</p>
</div>
<div id="connection-to-distance" class="section level3">
<h3><span class="header-section-number">13.7.5</span> Connection to distance</h3>
<p>The normal density is:</p>
<p><span class="math display">\[
p(x) = \frac{1}{\sqrt{2\pi} \sigma} \exp\left\{ - \frac{(x-\mu)^2}{\sigma^2}\right\}
\]</span></p>
<p>If we remove the constant <span class="math inline">\(1/(\sqrt{2\pi} \sigma)\)</span> and then take the log, we get:</p>
<p><span class="math display">\[
- \frac{(x-\mu)^2}{\sigma^2}
\]</span></p>
<p>which is the negative of a distance squared scaled by the standard deviation. For higher dimensions, the same is true except the scaling is more complex and involves correlations.</p>
</div>
</div>
<div id="case-study-more-than-three-classes" class="section level2">
<h2><span class="header-section-number">13.8</span> Case study: more than three classes</h2>
<p>We can generate an example with three categories like this:</p>
<div class="sourceCode" id="cb624"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb624-1" data-line-number="1"><span class="cf">if</span>(<span class="op">!</span><span class="kw">exists</span>(<span class="st">&quot;mnist&quot;</span>)) mnist &lt;-<span class="st"> </span><span class="kw">read_mnist</span>()</a>
<a class="sourceLine" id="cb624-2" data-line-number="2"><span class="kw">set.seed</span>(<span class="dv">3456</span>)</a>
<a class="sourceLine" id="cb624-3" data-line-number="3">index_<span class="dv">127</span> &lt;-<span class="st"> </span><span class="kw">sample</span>(<span class="kw">which</span>(mnist<span class="op">$</span>train<span class="op">$</span>labels <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">7</span>)), <span class="dv">2000</span>)</a>
<a class="sourceLine" id="cb624-4" data-line-number="4">y &lt;-<span class="st"> </span>mnist<span class="op">$</span>train<span class="op">$</span>labels[index_<span class="dv">127</span>] </a>
<a class="sourceLine" id="cb624-5" data-line-number="5">x &lt;-<span class="st"> </span>mnist<span class="op">$</span>train<span class="op">$</span>images[index_<span class="dv">127</span>,]</a>
<a class="sourceLine" id="cb624-6" data-line-number="6">index_train &lt;-<span class="st"> </span><span class="kw">createDataPartition</span>(y, <span class="dt">p=</span><span class="fl">0.8</span>, <span class="dt">list =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb624-7" data-line-number="7"><span class="co">## get the quadrants</span></a>
<a class="sourceLine" id="cb624-8" data-line-number="8">row_column &lt;-<span class="st"> </span><span class="kw">expand.grid</span>(<span class="dt">row=</span><span class="dv">1</span><span class="op">:</span><span class="dv">28</span>, <span class="dt">col=</span><span class="dv">1</span><span class="op">:</span><span class="dv">28</span>) </a>
<a class="sourceLine" id="cb624-9" data-line-number="9">upper_left_ind &lt;-<span class="st"> </span><span class="kw">which</span>(row_column<span class="op">$</span>col <span class="op">&lt;=</span><span class="st"> </span><span class="dv">14</span> <span class="op">&amp;</span><span class="st"> </span>row_column<span class="op">$</span>row <span class="op">&lt;=</span><span class="st"> </span><span class="dv">14</span>)</a>
<a class="sourceLine" id="cb624-10" data-line-number="10">lower_right_ind &lt;-<span class="st"> </span><span class="kw">which</span>(row_column<span class="op">$</span>col <span class="op">&gt;</span><span class="st"> </span><span class="dv">14</span> <span class="op">&amp;</span><span class="st"> </span>row_column<span class="op">$</span>row <span class="op">&gt;</span><span class="st"> </span><span class="dv">14</span>)</a>
<a class="sourceLine" id="cb624-11" data-line-number="11"><span class="co">## binarize the values. Above 200 is ink, below is no ink</span></a>
<a class="sourceLine" id="cb624-12" data-line-number="12">x &lt;-<span class="st"> </span>x <span class="op">&gt;</span><span class="st"> </span><span class="dv">200</span> </a>
<a class="sourceLine" id="cb624-13" data-line-number="13"><span class="co">## proportion of pixels in lower right quadrant</span></a>
<a class="sourceLine" id="cb624-14" data-line-number="14">x &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="kw">rowSums</span>(x[ ,upper_left_ind])<span class="op">/</span><span class="kw">rowSums</span>(x), </a>
<a class="sourceLine" id="cb624-15" data-line-number="15">           <span class="kw">rowSums</span>(x[ ,lower_right_ind])<span class="op">/</span><span class="kw">rowSums</span>(x)) </a>
<a class="sourceLine" id="cb624-16" data-line-number="16"><span class="co">##save data</span></a>
<a class="sourceLine" id="cb624-17" data-line-number="17">train_set &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">y =</span> <span class="kw">factor</span>(y[index_train]),</a>
<a class="sourceLine" id="cb624-18" data-line-number="18">                        <span class="dt">x_1 =</span> x[index_train,<span class="dv">1</span>], <span class="dt">x_2 =</span> x[index_train,<span class="dv">2</span>])</a>
<a class="sourceLine" id="cb624-19" data-line-number="19">test_set &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">y =</span> <span class="kw">factor</span>(y[<span class="op">-</span>index_train]),</a>
<a class="sourceLine" id="cb624-20" data-line-number="20">                       <span class="dt">x_1 =</span> x[<span class="op">-</span>index_train,<span class="dv">1</span>], <span class="dt">x_2 =</span> x[<span class="op">-</span>index_train,<span class="dv">2</span>])</a></code></pre></div>
<p>Here is the training data:</p>
<div class="sourceCode" id="cb625"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb625-1" data-line-number="1">train_set <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(x_<span class="dv">1</span>, x_<span class="dv">2</span>, <span class="dt">color=</span>y)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>()</a></code></pre></div>
<p><img src="Bayesian-Data-Science_files/figure-html/mnist-27-training-data-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>We can use the <strong>caret</strong> package to train the QDA model:</p>
<div class="sourceCode" id="cb626"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb626-1" data-line-number="1">train_qda &lt;-<span class="st"> </span><span class="kw">train</span>(y <span class="op">~</span><span class="st"> </span>., <span class="dt">method =</span> <span class="st">&quot;qda&quot;</span>, <span class="dt">data =</span> train_set)</a></code></pre></div>
<p>Now we estimate three conditional probabilities (although they have to add to 1):</p>
<div class="sourceCode" id="cb627"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb627-1" data-line-number="1"><span class="kw">predict</span>(train_qda, test_set, <span class="dt">type =</span> <span class="st">&quot;prob&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">head</span>()</a></code></pre></div>
<pre><code>       1       2       7
1 0.7655 0.23043 0.00405
2 0.2031 0.72514 0.07175
3 0.5396 0.45909 0.00132
4 0.0393 0.09419 0.86655
5 0.9600 0.00936 0.03063
6 0.9865 0.00724 0.00623</code></pre>
<p>Our predictions are one of the three classes:</p>
<div class="sourceCode" id="cb629"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb629-1" data-line-number="1"><span class="kw">predict</span>(train_qda, test_set) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">head</span>()</a></code></pre></div>
<pre><code>[1] 1 2 1 7 1 1
Levels: 1 2 7</code></pre>
<p>The confusion matrix is therefore a 3 by 3 table:</p>
<div class="sourceCode" id="cb631"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb631-1" data-line-number="1"><span class="kw">confusionMatrix</span>(<span class="kw">predict</span>(train_qda, test_set), test_set<span class="op">$</span>y)<span class="op">$</span>table</a></code></pre></div>
<pre><code>          Reference
Prediction   1   2   7
         1 111   9  11
         2  10  86  21
         7  21  28 102</code></pre>
<p>The accuracy is 0.749</p>
<p>Note that for sensitivity and specificity, we have a pair of values for <strong>each</strong> class. To define these terms, we need a binary outcome. We therefore have three columns: one for each class as the positives and the other two as the negatives.</p>
<p>To visualize what parts of the region are called 1, 2, and 7 we now need three colors:</p>
<p><img src="Bayesian-Data-Science_files/figure-html/three-classes-plot-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>The accuracy for LDA,
0.629,
is much worse because the model is more rigid. This is what the decision rule looks like:</p>
<p><img src="Bayesian-Data-Science_files/figure-html/lda-too-rigid-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>The results for kNN</p>
<div class="sourceCode" id="cb633"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb633-1" data-line-number="1">train_knn &lt;-<span class="st"> </span><span class="kw">train</span>(y <span class="op">~</span><span class="st"> </span>., <span class="dt">method =</span> <span class="st">&quot;knn&quot;</span>, <span class="dt">data =</span> train_set,</a>
<a class="sourceLine" id="cb633-2" data-line-number="2">                   <span class="dt">tuneGrid =</span> <span class="kw">data.frame</span>(<span class="dt">k =</span> <span class="kw">seq</span>(<span class="dv">15</span>, <span class="dv">51</span>, <span class="dv">2</span>)))</a></code></pre></div>
<p>are much better with an accuracy of
0.749. The decision rule looks like this:</p>
<p><img src="Bayesian-Data-Science_files/figure-html/three-classes-knn-better-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>Note that one of the limitations of generative models here is due to the lack of fit of the normal assumption, in particular for class 1.</p>
<div class="sourceCode" id="cb634"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb634-1" data-line-number="1">train_set <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">mutate</span>(<span class="dt">y =</span> <span class="kw">factor</span>(y)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb634-2" data-line-number="2"><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(x_<span class="dv">1</span>, x_<span class="dv">2</span>, <span class="dt">fill =</span> y, <span class="dt">color=</span>y)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb634-3" data-line-number="3"><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">show.legend =</span> <span class="ot">FALSE</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb634-4" data-line-number="4"><span class="st">  </span><span class="kw">stat_ellipse</span>(<span class="dt">type=</span><span class="st">&quot;norm&quot;</span>) </a></code></pre></div>
<p><img src="Bayesian-Data-Science_files/figure-html/three-classes-lack-of-fit-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>Generative models can be very powerful, but only when we are able to successfully approximate the joint distribution of predictors conditioned on each class.</p>
</div>
<div id="exercises-8" class="section level2">
<h2><span class="header-section-number">13.9</span> Exercises</h2>
<p>We are going to apply LDA and QDA to the <code>tissue_gene_expression</code> dataset. We will start with simple examples based on this dataset and then develop a realistic example.</p>
<p>1. Create a dataset with just the classes “cerebellum” and “hippocampus” (two parts of the brain) and a predictor matrix with 10 randomly selected columns.</p>
<div class="sourceCode" id="cb635"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb635-1" data-line-number="1"><span class="kw">set.seed</span>(<span class="dv">1993</span>)</a>
<a class="sourceLine" id="cb635-2" data-line-number="2"><span class="kw">data</span>(<span class="st">&quot;tissue_gene_expression&quot;</span>)</a>
<a class="sourceLine" id="cb635-3" data-line-number="3">tissues &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;cerebellum&quot;</span>, <span class="st">&quot;hippocampus&quot;</span>)</a>
<a class="sourceLine" id="cb635-4" data-line-number="4">ind &lt;-<span class="st"> </span><span class="kw">which</span>(tissue_gene_expression<span class="op">$</span>y <span class="op">%in%</span><span class="st"> </span>tissues)</a>
<a class="sourceLine" id="cb635-5" data-line-number="5">y &lt;-<span class="st"> </span><span class="kw">droplevels</span>(tissue_gene_expression<span class="op">$</span>y[ind])</a>
<a class="sourceLine" id="cb635-6" data-line-number="6">x &lt;-<span class="st"> </span>tissue_gene_expression<span class="op">$</span>x[ind, ]</a>
<a class="sourceLine" id="cb635-7" data-line-number="7">x &lt;-<span class="st"> </span>x[, <span class="kw">sample</span>(<span class="kw">ncol</span>(x), <span class="dv">10</span>)]</a></code></pre></div>
<p>Use the <code>train</code> function to estimate the accuracy of LDA.</p>
<p>2. In this case, LDA fits two 10-dimensional normal distributions. Look at the fitted model by looking at the <code>finalModel</code> component of the result of train. Notice there is a component called <code>means</code> that includes the estimate <code>means</code> of both distributions. Plot the mean vectors against each other and determine which predictors (genes) appear to be driving the algorithm.</p>
<p>3. Repeat exercises 1 with QDA. Does it have a higher accuracy than LDA?</p>
<p>4. Are the same predictors (genes) driving the algorithm? Make a plot as in exercise 2.</p>
<p>5. One thing we see in the previous plot is that the value of predictors correlate in both groups: some predictors are low in both groups while others are high in both groups. The mean value of each predictor, <code>colMeans(x)</code>, is not informative or useful for prediction, and often for interpretation purposes it is useful to center or scale each column. This can be achieved with the <code>preProcessing</code> argument in <code>train</code>. Re-run LDA with <code>preProcessing = &quot;scale&quot;</code>. Note that accuracy does not change but see how it is easier to identify the predictors that differ more between groups in the plot made in exercise 4.</p>
<p>6. In the previous exercises we saw that both approaches worked well. Plot the predictor values for the two genes with the largest differences between the two groups in a scatterplot to see how they appear to follow a bivariate distribution as assumed by the LDA and QDA approaches. Color the points by the outcome.</p>
<p>7. Now we are going to increase the complexity of the challenge slightly: we will consider all the tissue types.</p>
<div class="sourceCode" id="cb636"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb636-1" data-line-number="1"><span class="kw">set.seed</span>(<span class="dv">1993</span>)</a>
<a class="sourceLine" id="cb636-2" data-line-number="2"><span class="kw">data</span>(<span class="st">&quot;tissue_gene_expression&quot;</span>)</a>
<a class="sourceLine" id="cb636-3" data-line-number="3">y &lt;-<span class="st"> </span>tissue_gene_expression<span class="op">$</span>y</a>
<a class="sourceLine" id="cb636-4" data-line-number="4">x &lt;-<span class="st"> </span>tissue_gene_expression<span class="op">$</span>x</a>
<a class="sourceLine" id="cb636-5" data-line-number="5">x &lt;-<span class="st"> </span>x[, <span class="kw">sample</span>(<span class="kw">ncol</span>(x), <span class="dv">10</span>)]</a></code></pre></div>
<p>What accuracy do you get with LDA?</p>
<p>8. We see that the results are slightly worse. Use the <code>confusionMatrix</code> function to learn what type of errors we are making.</p>
<p>9. Plot an image of the centers of the seven 10-dimensional normal distributions.</p>

</div>
<div id="classification-and-regression-trees-cart" class="section level2">
<h2><span class="header-section-number">13.10</span> Classification and regression trees (CART)</h2>
<div id="the-curse-of-dimensionality" class="section level3">
<h3><span class="header-section-number">13.10.1</span> The curse of dimensionality</h3>
<p>We described how methods such as LDA and QDA are not meant to be used with many predictors <span class="math inline">\(p\)</span> because the number of parameters that we need to estimate becomes too large. For example, with the digits example <span class="math inline">\(p=784\)</span>, we would have over 600,000 parameters with LDA, and we would multiply that by the number of classes for QDA. Kernel methods such as kNN or local regression do not have model parameters to estimate. However, they also face a challenge when multiple predictors are used due to what is referred to as the <em>curse of dimensionality</em>. The <em>dimension</em> here refers to the fact that when we have <span class="math inline">\(p\)</span> predictors, the distance between two observations is computed in <span class="math inline">\(p\)</span>-dimensional space.</p>
<p>A useful way of understanding the curse of dimensionality is by considering how large we have to make a span/neighborhood/window to include a given percentage of the data. Remember that with larger neighborhoods, our methods lose flexibility.</p>
<p>For example, suppose we have one continuous predictor with equally spaced points in the [0,1] interval and we want to create windows that include 1/10th of data. Then it’s easy to see that our windows have to be of size 0.1:</p>
<p><img src="Bayesian-Data-Science_files/figure-html/curse-of-dim-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p>Now, for two predictors, if we decide to keep the neighborhood just as small, 10% for each dimension, we include only 1 point. If we want to include 10% of the data, then we need to increase the size of each side of the square to <span class="math inline">\(\sqrt{.10} \approx .316\)</span>:</p>
<p><img src="Bayesian-Data-Science_files/figure-html/curse-of-dim-2-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p>Using the same logic, if we want to include 10% of the data in a three-dimensional space, then the side of each cube is <span class="math inline">\(\sqrt[3]{.10} \approx 0.464\)</span>.
In general, to include 10% of the data in a case with <span class="math inline">\(p\)</span> dimensions, we need an interval with each side of size <span class="math inline">\(\sqrt[p]{.10}\)</span> of the total. This proportion gets close to 1 quickly, and if the proportion is 1 it means we include all the data and are no longer smoothing.</p>
<div class="sourceCode" id="cb637"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb637-1" data-line-number="1"><span class="kw">library</span>(tidyverse)</a>
<a class="sourceLine" id="cb637-2" data-line-number="2">p &lt;-<span class="st"> </span><span class="dv">1</span><span class="op">:</span><span class="dv">100</span></a>
<a class="sourceLine" id="cb637-3" data-line-number="3"><span class="kw">qplot</span>(p, <span class="fl">.1</span><span class="op">^</span>(<span class="dv">1</span><span class="op">/</span>p), <span class="dt">ylim =</span> <span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>))</a></code></pre></div>
<p><img src="Bayesian-Data-Science_files/figure-html/curse-of-dim-4-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>By the time we reach 100 predictors, the neighborhood is no longer very local, as each side covers almost the entire dataset.</p>
<p>Here we look at a set of elegant and versatile methods that adapt to higher dimensions and also allow these regions to take more complex shapes while still producing models that are interpretable. These are very popular, well-known and studied methods. We will concentrate on regression and decision trees and their extension to random forests.</p>
</div>
<div id="cart-motivation" class="section level3">
<h3><span class="header-section-number">13.10.2</span> CART motivation</h3>
<p>To motivate this section, we will use a new dataset
that includes the breakdown of the composition of olive oil into 8 fatty acids:</p>
<div class="sourceCode" id="cb638"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb638-1" data-line-number="1"><span class="kw">library</span>(tidyverse)</a>
<a class="sourceLine" id="cb638-2" data-line-number="2"><span class="kw">library</span>(dslabs)</a>
<a class="sourceLine" id="cb638-3" data-line-number="3"><span class="kw">data</span>(<span class="st">&quot;olive&quot;</span>)</a>
<a class="sourceLine" id="cb638-4" data-line-number="4"><span class="kw">names</span>(olive)</a></code></pre></div>
<pre><code> [1] &quot;region&quot;      &quot;area&quot;        &quot;palmitic&quot;    &quot;palmitoleic&quot; &quot;stearic&quot;    
 [6] &quot;oleic&quot;       &quot;linoleic&quot;    &quot;linolenic&quot;   &quot;arachidic&quot;   &quot;eicosenoic&quot; </code></pre>
<p>For illustrative purposes, we will try to predict the region using the fatty acid composition values as predictors.</p>
<div class="sourceCode" id="cb640"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb640-1" data-line-number="1"><span class="kw">table</span>(olive<span class="op">$</span>region)</a></code></pre></div>
<pre><code>
Northern Italy       Sardinia Southern Italy 
           151             98            323 </code></pre>
<p>We remove the <code>area</code> column because we won’t use it as a predictor.</p>
<div class="sourceCode" id="cb642"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb642-1" data-line-number="1">olive &lt;-<span class="st"> </span><span class="kw">select</span>(olive, <span class="op">-</span>area)</a></code></pre></div>
<p>Let’s very quickly try to predict the region using kNN:</p>
<div class="sourceCode" id="cb643"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb643-1" data-line-number="1"><span class="kw">library</span>(caret)</a>
<a class="sourceLine" id="cb643-2" data-line-number="2">fit &lt;-<span class="st"> </span><span class="kw">train</span>(region <span class="op">~</span><span class="st"> </span>.,  <span class="dt">method =</span> <span class="st">&quot;knn&quot;</span>, </a>
<a class="sourceLine" id="cb643-3" data-line-number="3">             <span class="dt">tuneGrid =</span> <span class="kw">data.frame</span>(<span class="dt">k =</span> <span class="kw">seq</span>(<span class="dv">1</span>, <span class="dv">15</span>, <span class="dv">2</span>)), </a>
<a class="sourceLine" id="cb643-4" data-line-number="4">             <span class="dt">data =</span> olive)</a>
<a class="sourceLine" id="cb643-5" data-line-number="5"><span class="kw">ggplot</span>(fit)</a></code></pre></div>
<p><img src="Bayesian-Data-Science_files/figure-html/olive-knn-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>We see that using just one neighbor, we can predict relatively well. However, a bit of data exploration reveals that we should be able to do even better. For example, if we look at the distribution of each predictor stratified by region we see that eicosenoic is only present in Southern Italy and that linoleic separates Northern Italy from Sardinia.</p>
<div class="sourceCode" id="cb644"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb644-1" data-line-number="1">olive <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">gather</span>(fatty_acid, percentage, <span class="op">-</span>region) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb644-2" data-line-number="2"><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(region, percentage, <span class="dt">fill =</span> region)) <span class="op">+</span></a>
<a class="sourceLine" id="cb644-3" data-line-number="3"><span class="st">  </span><span class="kw">geom_boxplot</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb644-4" data-line-number="4"><span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span>fatty_acid, <span class="dt">scales =</span> <span class="st">&quot;free&quot;</span>, <span class="dt">ncol =</span> <span class="dv">4</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb644-5" data-line-number="5"><span class="st">  </span><span class="kw">theme</span>(<span class="dt">axis.text.x =</span> <span class="kw">element_blank</span>(), <span class="dt">legend.position=</span><span class="st">&quot;bottom&quot;</span>)</a></code></pre></div>
<p><img src="Bayesian-Data-Science_files/figure-html/olive-eda-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>This implies that we should be able to build an algorithm that predicts perfectly! We can see this clearly by plotting the values for eicosenoic and linoleic.</p>
<div class="sourceCode" id="cb645"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb645-1" data-line-number="1">olive <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb645-2" data-line-number="2"><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(eicosenoic, linoleic, <span class="dt">color =</span> region)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb645-3" data-line-number="3"><span class="st">  </span><span class="kw">geom_point</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb645-4" data-line-number="4"><span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> <span class="fl">0.065</span>, <span class="dt">lty =</span> <span class="dv">2</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb645-5" data-line-number="5"><span class="st">  </span><span class="kw">geom_segment</span>(<span class="dt">x =</span> <span class="fl">-0.2</span>, <span class="dt">y =</span> <span class="fl">10.54</span>, <span class="dt">xend =</span> <span class="fl">0.065</span>, <span class="dt">yend =</span> <span class="fl">10.54</span>, </a>
<a class="sourceLine" id="cb645-6" data-line-number="6">               <span class="dt">color =</span> <span class="st">&quot;black&quot;</span>, <span class="dt">lty =</span> <span class="dv">2</span>)</a></code></pre></div>
<p><img src="Bayesian-Data-Science_files/figure-html/olive-two-predictors-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>In Section <a href="#predictor-space"><strong>??</strong></a> we define predictor spaces. The predictor space here consists of eight-dimensional points with values between 0 and 100. In the plot above, we show the space defined by the two predictors eicosenoic and linoleic, and, by eye,
we can construct a prediction rule that partitions the predictor space so that each partition contains only outcomes of a one category. This in turn can be used to define an algorithm with perfect accuracy. Specifically, we define the following decision rule. If eicosenoic is larger than 0.065, predict Southern Italy. If not, then if linoleic is larger than <span class="math inline">\(10.535\)</span>, predict Sardinia, and if lower, predict Northern Italy. We can draw this decision tree like this:</p>
<p><img src="Bayesian-Data-Science_files/figure-html/olive-tree-1.png" width="50%" style="display: block; margin: auto;" /></p>
<p>Decision trees like this are often used in practice. For example, to decide on a person’s risk of poor outcome after having a heart attack, doctors use the following:</p>
<p><img src="ml/img/Decision-Tree-for-Heart-Attack-Victim-adapted-from-Gigerenzer-et-al-1999-4.png" width="50%" style="display: block; margin: auto;" /></p>
<p>(Source: Walton 2010 Informal Logic, Vol. 30, No. 2, pp. 159-184<a href="#fn17" class="footnote-ref" id="fnref17"><sup>17</sup></a>.)</p>
<p>A tree is basically a flow chart of yes or no questions. The general idea of the methods we are describing is to define an algorithm that uses data to create these trees with predictions at the ends, referred to as <em>nodes</em>.
Regression and decision trees operate by predicting an outcome variable <span class="math inline">\(Y\)</span> by partitioning the predictors.</p>
</div>
<div id="regression-trees" class="section level3">
<h3><span class="header-section-number">13.10.3</span> Regression trees</h3>
<p>When the outcome is continuous, we call the method a <em>regression</em> tree. To introduce regression trees, we will use the 2008 poll data used in previous sections to describe the basic idea of how we build these algorithms. As with other machine learning algorithms, we will try to estimate the conditional expectation <span class="math inline">\(f(x) = \mbox{E}(Y | X = x)\)</span> with <span class="math inline">\(Y\)</span> the poll margin and <span class="math inline">\(x\)</span> the day.</p>
<div class="sourceCode" id="cb646"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb646-1" data-line-number="1"><span class="kw">data</span>(<span class="st">&quot;polls_2008&quot;</span>)</a>
<a class="sourceLine" id="cb646-2" data-line-number="2"><span class="kw">qplot</span>(day, margin, <span class="dt">data =</span> polls_<span class="dv">2008</span>)</a></code></pre></div>
<p><img src="Bayesian-Data-Science_files/figure-html/polls-2008-again-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>The general idea here is to build a decision tree and, at the end of each <em>node</em>, obtain a predictor <span class="math inline">\(\hat{y}\)</span>. A mathematical way to describe this is to say that we are partitioning the predictor space into <span class="math inline">\(J\)</span> non-overlapping regions, <span class="math inline">\(R_1, R_2, \ldots, R_J\)</span>, and then for any predictor <span class="math inline">\(x\)</span> that falls within region <span class="math inline">\(R_j\)</span>, estimate <span class="math inline">\(f(x)\)</span> with the average of the training observations <span class="math inline">\(y_i\)</span> for which the associated predictor <span class="math inline">\(x_i\)</span> is also in <span class="math inline">\(R_j\)</span>.</p>
<p>But how do we decide on the partition <span class="math inline">\(R_1, R_2, \ldots, R_J\)</span> and how do we choose <span class="math inline">\(J\)</span>? Here is where the algorithm gets a bit complicated.</p>
<p>Regression trees create partitions recursively. We start the algorithm with one partition, the entire predictor space. In our simple first example, this space is the interval [-155, 1]. But after the first step we will have two partitions. After the second step we will split one of these partitions into two and will have three partitions, then four, then five, and so on. We describe how we pick the partition to further partition, and when to stop, later.</p>
<p>Once we select a partition <span class="math inline">\(\mathbf{x}\)</span> to split in order to create the new partitions, we find a predictor <span class="math inline">\(j\)</span> and value <span class="math inline">\(s\)</span> that define two new partitions, which we will call <span class="math inline">\(R_1(j,s)\)</span> and <span class="math inline">\(R_2(j,s)\)</span>, that split our observations in the current partition by asking if <span class="math inline">\(x_j\)</span> is bigger than <span class="math inline">\(s\)</span>:</p>
<p><span class="math display">\[
R_1(j,s) = \{\mathbf{x} \mid x_j &lt; s\} \mbox{  and  } R_2(j,s) = \{\mathbf{x} \mid x_j \geq s\}
\]</span></p>
<p>In our current example we only have one predictor, so we will always choose <span class="math inline">\(j=1\)</span>, but in general this will not be the case. Now, after we define the new partitions <span class="math inline">\(R_1\)</span> and <span class="math inline">\(R_2\)</span>, and we decide to stop the partitioning, we compute predictors by taking the average of all the observations <span class="math inline">\(y\)</span> for which the associated <span class="math inline">\(\mathbf{x}\)</span> is in <span class="math inline">\(R_1\)</span> and <span class="math inline">\(R_2\)</span>. We refer to these two as <span class="math inline">\(\hat{y}_{R_1}\)</span> and <span class="math inline">\(\hat{y}_{R_2}\)</span> respectively.</p>
<p>But how do we pick <span class="math inline">\(j\)</span> and <span class="math inline">\(s\)</span>? Basically we find the pair that minimizes the residual sum of square (RSS):
<span class="math display">\[
\sum_{i:\, x_i \in R_1(j,s)} (y_i - \hat{y}_{R_1})^2 +
\sum_{i:\, x_i \in R_2(j,s)} (y_i - \hat{y}_{R_2})^2
\]</span></p>
<p>This is then applied recursively to the new regions <span class="math inline">\(R_1\)</span> and <span class="math inline">\(R_2\)</span>. We describe how we stop later, but once we are done partitioning the predictor space into regions, in each region a prediction is made using the observations in that region.</p>
<p>Let’s take a look at what this algorithm does on the 2008 presidential election poll data. We will use the <code>rpart</code> function in the <strong>rpart</strong> package.</p>
<div class="sourceCode" id="cb647"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb647-1" data-line-number="1"><span class="kw">library</span>(rpart)</a>
<a class="sourceLine" id="cb647-2" data-line-number="2">fit &lt;-<span class="st"> </span><span class="kw">rpart</span>(margin <span class="op">~</span><span class="st"> </span>., <span class="dt">data =</span> polls_<span class="dv">2008</span>)</a></code></pre></div>
<p>Here, there is only one predictor. Thus we do not have to decide which predictor <span class="math inline">\(j\)</span> to split by, we simply have to decide what value <span class="math inline">\(s\)</span> we use to split. We can visually see where the splits were made:</p>
<div class="sourceCode" id="cb648"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb648-1" data-line-number="1"><span class="kw">plot</span>(fit, <span class="dt">margin =</span> <span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb648-2" data-line-number="2"><span class="kw">text</span>(fit, <span class="dt">cex =</span> <span class="fl">0.75</span>)</a></code></pre></div>
<p><img src="Bayesian-Data-Science_files/figure-html/polls-2008-tree-1.png" width="60%" style="display: block; margin: auto;" /></p>
<p>The first split is made on day 39.5. One of those regions is then split at day 86.5. The two resulting new partitions are split on days 49.5 and 117.5, respectively, and so on. We end up with 8 partitions. The final estimate <span class="math inline">\(\hat{f}(x)\)</span> looks like this:</p>
<div class="sourceCode" id="cb649"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb649-1" data-line-number="1">polls_<span class="dv">2008</span> <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb649-2" data-line-number="2"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">y_hat =</span> <span class="kw">predict</span>(fit)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb649-3" data-line-number="3"><span class="st">  </span><span class="kw">ggplot</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb649-4" data-line-number="4"><span class="st">  </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(day, margin)) <span class="op">+</span></a>
<a class="sourceLine" id="cb649-5" data-line-number="5"><span class="st">  </span><span class="kw">geom_step</span>(<span class="kw">aes</span>(day, y_hat), <span class="dt">col=</span><span class="st">&quot;red&quot;</span>)</a></code></pre></div>
<p><img src="Bayesian-Data-Science_files/figure-html/polls-2008-tree-fit-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>Note that the algorithm stopped partitioning at 8. Now we explain how this decision is made.</p>
<p>First we need to define the term <em>complexity parameter</em> (cp). Every time we split and define two new partitions, our training set RSS decreases. This is because with more partitions, our model has more flexibility to adapt to the training data. In fact, if you split until every point is its own partition, then RSS goes all the way down to 0 since the average of one value is that same value. To avoid this, the algorithm sets a minimum for how much the RSS must improve for another partition to be added. This parameter is referred to as the <em>complexity parameter</em> (cp). The RSS must improve by a factor of cp for the new partition to be added. Large values of cp will therefore force the algorithm to stop earlier which results in fewer nodes.</p>
<p>However, cp is not the only parameter used to decide if we should partition a current partition or not. Another common parameter is the minimum number of observations required in a partition before partitioning it further. The argument used in the <code>rpart</code> function is <code>minsplit</code> and the default is 20. The <code>rpart</code> implementation of regression trees also permits users to determine a minimum number of observations in each node. The argument is <code>minbucket</code> and defaults to <code>round(minsplit/3)</code>.</p>
<p>As expected, if we set <code>cp = 0</code> and <code>minsplit = 2</code>, then our prediction is as flexible as possible and our predictor is our original data:</p>
<div class="sourceCode" id="cb650"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb650-1" data-line-number="1">fit &lt;-<span class="st"> </span><span class="kw">rpart</span>(margin <span class="op">~</span><span class="st"> </span>., <span class="dt">data =</span> polls_<span class="dv">2008</span>, </a>
<a class="sourceLine" id="cb650-2" data-line-number="2">             <span class="dt">control =</span> <span class="kw">rpart.control</span>(<span class="dt">cp =</span> <span class="dv">0</span>, <span class="dt">minsplit =</span> <span class="dv">2</span>))</a>
<a class="sourceLine" id="cb650-3" data-line-number="3">polls_<span class="dv">2008</span> <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb650-4" data-line-number="4"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">y_hat =</span> <span class="kw">predict</span>(fit)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb650-5" data-line-number="5"><span class="st">  </span><span class="kw">ggplot</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb650-6" data-line-number="6"><span class="st">  </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(day, margin)) <span class="op">+</span></a>
<a class="sourceLine" id="cb650-7" data-line-number="7"><span class="st">  </span><span class="kw">geom_step</span>(<span class="kw">aes</span>(day, y_hat), <span class="dt">col=</span><span class="st">&quot;red&quot;</span>)</a></code></pre></div>
<p><img src="Bayesian-Data-Science_files/figure-html/polls-2008-tree-over-fit-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>Intuitively we know that this is not a good approach as it will generally result in over-training. These <code>cp</code>, <code>minsplit</code>, and <code>minbucket</code>, three parameters can be used to control the variability of the final predictors. The larger these values are the more data is averaged to compute a predictor and thus reduce variability. The drawback is that it restricts flexibility.</p>
<p>So how do we pick these parameters? We can use cross validation, described in Chapter <a href="12-cross-validation.html#cross-validation">12</a>, just like with any tuning parameter. Here is an example of using cross validation to chose cp.</p>
<div class="sourceCode" id="cb651"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb651-1" data-line-number="1"><span class="kw">library</span>(caret)</a>
<a class="sourceLine" id="cb651-2" data-line-number="2">train_rpart &lt;-<span class="st"> </span><span class="kw">train</span>(margin <span class="op">~</span><span class="st"> </span>., </a>
<a class="sourceLine" id="cb651-3" data-line-number="3">                     <span class="dt">method =</span> <span class="st">&quot;rpart&quot;</span>,</a>
<a class="sourceLine" id="cb651-4" data-line-number="4">                     <span class="dt">tuneGrid =</span> <span class="kw">data.frame</span>(<span class="dt">cp =</span> <span class="kw">seq</span>(<span class="dv">0</span>, <span class="fl">0.05</span>, <span class="dt">len =</span> <span class="dv">25</span>)),</a>
<a class="sourceLine" id="cb651-5" data-line-number="5">                     <span class="dt">data =</span> polls_<span class="dv">2008</span>)</a>
<a class="sourceLine" id="cb651-6" data-line-number="6"><span class="kw">ggplot</span>(train_rpart)</a></code></pre></div>
<p><img src="Bayesian-Data-Science_files/figure-html/polls-2008-tree-train-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>To see the resulting tree, we access the <code>finalModel</code> and plot it:</p>
<div class="sourceCode" id="cb652"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb652-1" data-line-number="1"><span class="kw">plot</span>(train_rpart<span class="op">$</span>finalModel, <span class="dt">margin =</span> <span class="fl">0.1</span>)</a>
<a class="sourceLine" id="cb652-2" data-line-number="2"><span class="kw">text</span>(train_rpart<span class="op">$</span>finalModel, <span class="dt">cex =</span> <span class="fl">0.75</span>)</a></code></pre></div>
<p><img src="Bayesian-Data-Science_files/figure-html/polls-2008-final-model-1.png" width="80%" style="display: block; margin: auto;" /></p>
<p>And because we only have one predictor, we can actually plot <span class="math inline">\(\hat{f}(x)\)</span>:</p>
<div class="sourceCode" id="cb653"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb653-1" data-line-number="1">polls_<span class="dv">2008</span> <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb653-2" data-line-number="2"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">y_hat =</span> <span class="kw">predict</span>(train_rpart)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb653-3" data-line-number="3"><span class="st">  </span><span class="kw">ggplot</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb653-4" data-line-number="4"><span class="st">  </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(day, margin)) <span class="op">+</span></a>
<a class="sourceLine" id="cb653-5" data-line-number="5"><span class="st">  </span><span class="kw">geom_step</span>(<span class="kw">aes</span>(day, y_hat), <span class="dt">col=</span><span class="st">&quot;red&quot;</span>)</a></code></pre></div>
<p><img src="Bayesian-Data-Science_files/figure-html/polls-2008-final-fit-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>Note that if we already have a tree and want to apply a higher cp value, we can use the <code>prune</code> function. We call this <em>pruning</em> a tree because we are snipping off partitions that do not meet a <code>cp</code> criterion. We previously created a tree that used a <code>cp = 0</code> and saved it to <code>fit</code>. We can prune it like this:</p>
<div class="sourceCode" id="cb654"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb654-1" data-line-number="1">pruned_fit &lt;-<span class="st"> </span><span class="kw">prune</span>(fit, <span class="dt">cp =</span> <span class="fl">0.01</span>)</a></code></pre></div>
</div>
<div id="classification-decision-trees" class="section level3">
<h3><span class="header-section-number">13.10.4</span> Classification (decision) trees</h3>
<p>Classification trees, or decision trees, are used in prediction problems where the outcome is categorical. We use the same partitioning principle with some differences to account for the fact that we are now working with a categorical outcome.</p>
<p>The first difference is that we form predictions by calculating which class is the most common among the training set observations within the partition, rather than taking the average in each partition (as we can’t take the average of categories).</p>
<p>The second is that we can no longer use RSS to choose the partition. While we could use the naive approach of looking for partitions that minimize training error, better performing approaches use more sophisticated metrics. Two of the more popular ones are the <em>Gini Index</em> and <em>Entropy</em>.</p>
<p>In a perfect scenario, the outcomes in each of our partitions are all of the same category since this will permit perfect accuracy. The <em>Gini Index</em> is going to be 0 in this scenario, and become larger the more we deviate from this scenario. To define the Gini Index, we define <span class="math inline">\(\hat{p}_{j,k}\)</span> as the proportion of observations in partition <span class="math inline">\(j\)</span> that are of class <span class="math inline">\(k\)</span>. The Gini Index is defined as</p>
<p><span class="math display">\[
\mbox{Gini}(j) = \sum_{k=1}^K \hat{p}_{j,k}(1-\hat{p}_{j,k})
\]</span></p>
<p>If you study the formula carefully you will see that it is in fact 0 in the perfect scenario described above.</p>
<p><em>Entropy</em> is a very similar quantity, defined as</p>
<p><span class="math display">\[
\mbox{entropy}(j) = -\sum_{k=1}^K \hat{p}_{j,k}\log(\hat{p}_{j,k}), \mbox{ with } 0 \times \log(0) \mbox{ defined as }0
\]</span></p>
<p>Let us look at how a classification tree performs on the digits example we examined before:</p>
<p>We can use this code to run the algorithm and plot the resulting tree:</p>
<div class="sourceCode" id="cb655"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb655-1" data-line-number="1">train_rpart &lt;-<span class="st"> </span><span class="kw">train</span>(y <span class="op">~</span><span class="st"> </span>.,</a>
<a class="sourceLine" id="cb655-2" data-line-number="2">                     <span class="dt">method =</span> <span class="st">&quot;rpart&quot;</span>,</a>
<a class="sourceLine" id="cb655-3" data-line-number="3">                     <span class="dt">tuneGrid =</span> <span class="kw">data.frame</span>(<span class="dt">cp =</span> <span class="kw">seq</span>(<span class="fl">0.0</span>, <span class="fl">0.1</span>, <span class="dt">len =</span> <span class="dv">25</span>)),</a>
<a class="sourceLine" id="cb655-4" data-line-number="4">                     <span class="dt">data =</span> mnist_<span class="dv">27</span><span class="op">$</span>train)</a>
<a class="sourceLine" id="cb655-5" data-line-number="5"><span class="kw">plot</span>(train_rpart)</a></code></pre></div>
<p><img src="Bayesian-Data-Science_files/figure-html/mnist-27-tree-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>The accuracy achieved by this approach is better than what we got with regression, but is not as good as what we achieved with kernel methods:</p>
<div class="sourceCode" id="cb656"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb656-1" data-line-number="1">y_hat &lt;-<span class="st"> </span><span class="kw">predict</span>(train_rpart, mnist_<span class="dv">27</span><span class="op">$</span>test)</a>
<a class="sourceLine" id="cb656-2" data-line-number="2"><span class="kw">confusionMatrix</span>(y_hat, mnist_<span class="dv">27</span><span class="op">$</span>test<span class="op">$</span>y)<span class="op">$</span>overall[<span class="st">&quot;Accuracy&quot;</span>]</a></code></pre></div>
<pre><code>Accuracy 
    0.82 </code></pre>
<p>The plot of the estimated conditional probability shows us the limitations of classification trees:</p>
<p><img src="Bayesian-Data-Science_files/figure-html/rf-cond-prob-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>Note that with decision trees, it is difficult to make the boundaries smooth since each partition creates a discontinuity.</p>
<p>Classification trees have certain advantages that make them very useful. They are highly interpretable, even more so than linear models. They are easy to visualize (if small enough). Finally, they can model human decision processes and don’t require use of dummy predictors for categorical variables. On the other hand, the approach via recursive partitioning can easily over-train and is therefore a bit harder to train than, for example, linear regression or kNN. Furthermore, in terms of accuracy, it is rarely the best performing method since it is not very flexible and is highly unstable to changes in training data. Random forests, explained next, improve on several of these shortcomings.</p>
</div>
</div>
<div id="random-forests" class="section level2">
<h2><span class="header-section-number">13.11</span> Random forests</h2>
<p>Random forests are a <strong>very popular</strong> machine learning approach that addresses the shortcomings of decision trees using a clever idea. The goal is to improve prediction performance and reduce instability by <em>averaging</em> multiple decision trees (a forest of trees constructed with randomness). It has two features that help accomplish this.</p>
<p>The first step is <em>bootstrap aggregation</em> or <em>bagging</em>. The general idea is to generate many predictors, each using regression or classification trees, and then forming a final prediction based on the average prediction of all these trees. To assure that the individual trees are not the same, we use the bootstrap to induce randomness. These two features combined explain the name: the bootstrap makes the individual trees <strong>randomly</strong> different, and the combination of trees is the <strong>forest</strong>. The specific steps are as follows.</p>
<p>1. Build <span class="math inline">\(B\)</span> decision trees using the training set. We refer to the fitted models as <span class="math inline">\(T_1, T_2, \dots, T_B\)</span>. We later explain how we ensure they are different.</p>
<p>2. For every observation in the test set, form a prediction <span class="math inline">\(\hat{y}_j\)</span> using tree <span class="math inline">\(T_j\)</span>.</p>
<p>3. For continuous outcomes, form a final prediction with the average <span class="math inline">\(\hat{y} = \frac{1}{B} \sum_{j=1}^B \hat{y}_j\)</span>. For categorical data classification, predict <span class="math inline">\(\hat{y}\)</span> with majority vote (most frequent class among <span class="math inline">\(\hat{y}_1, \dots, \hat{y}_T\)</span>).</p>
<p>So how do we get different decision trees from a single training set? For this, we use randomness in two ways which we explain in the steps below. Let <span class="math inline">\(N\)</span> be the number of observations in the training set. To create <span class="math inline">\(T_j, \, j=1,\ldots,B\)</span> from the training set we do the following:</p>
<p>1. Create a bootstrap training set by sampling <span class="math inline">\(N\)</span> observations from the training set <strong>with replacement</strong>. This is the first way to induce randomness.</p>
<p>2. A large number of features is typical in machine learning challenges. Often, many features can be informative but including them all in the model may result in overfitting. The second way random forests induce randomness is by randomly selecting features to be included in the building of each tree. A different random subset is selected for each tree. This reduces correlation between trees in the forest, thereby improving prediction accuracy.</p>
<p>To illustrate how the first steps can result in smoother estimates we will demonstrate by fitting a random forest to
the 2008 polls data. We will use the <code>randomForest</code> function in the <strong>randomForest</strong> package:</p>
<div class="sourceCode" id="cb658"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb658-1" data-line-number="1"><span class="kw">library</span>(randomForest)</a>
<a class="sourceLine" id="cb658-2" data-line-number="2">fit &lt;-<span class="st"> </span><span class="kw">randomForest</span>(margin<span class="op">~</span>., <span class="dt">data =</span> polls_<span class="dv">2008</span>) </a></code></pre></div>
<p>Note that if we apply the function <code>plot</code> to the resulting object, stored in <code>fit</code>, we see how the error rate of our algorithm changes as we add trees.</p>
<div class="sourceCode" id="cb659"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb659-1" data-line-number="1">rafalib<span class="op">::</span><span class="kw">mypar</span>()</a>
<a class="sourceLine" id="cb659-2" data-line-number="2"><span class="kw">plot</span>(fit)</a></code></pre></div>
<p><img src="Bayesian-Data-Science_files/figure-html/more-trees-better-fit-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>We can see that in this case, the accuracy improves as we add more trees until about 30 trees where accuracy stabilizes.</p>
<p>The resulting estimate for this random forest can be seen like this:</p>
<div class="sourceCode" id="cb660"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb660-1" data-line-number="1">polls_<span class="dv">2008</span> <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb660-2" data-line-number="2"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">y_hat =</span> <span class="kw">predict</span>(fit, <span class="dt">newdata =</span> polls_<span class="dv">2008</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb660-3" data-line-number="3"><span class="st">  </span><span class="kw">ggplot</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb660-4" data-line-number="4"><span class="st">  </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(day, margin)) <span class="op">+</span></a>
<a class="sourceLine" id="cb660-5" data-line-number="5"><span class="st">  </span><span class="kw">geom_line</span>(<span class="kw">aes</span>(day, y_hat), <span class="dt">col=</span><span class="st">&quot;red&quot;</span>)</a></code></pre></div>
<p><img src="Bayesian-Data-Science_files/figure-html/polls-2008-rf-fit-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>Notice that the random forest estimate is much smoother than what we achieved with the regression tree in the previous section. This is possible because the average of many step functions can be smooth. We can see this by visually examining how the estimate changes as we add more trees. In the following figure you see each of the bootstrap samples for several values of <span class="math inline">\(b\)</span> and for each one we see the tree that is fitted in grey, the previous trees that were fitted in lighter grey, and the result of averaging all the trees estimated up to that point.</p>
<p><img src="ml/img/rf.gif" width="100%" style="display: block; margin: auto;" /></p>
<p>Here is the random forest fit for our digits example based on two predictors:</p>
<div class="sourceCode" id="cb661"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb661-1" data-line-number="1"><span class="kw">library</span>(randomForest)</a>
<a class="sourceLine" id="cb661-2" data-line-number="2">train_rf &lt;-<span class="st"> </span><span class="kw">randomForest</span>(y <span class="op">~</span><span class="st"> </span>., <span class="dt">data=</span>mnist_<span class="dv">27</span><span class="op">$</span>train)</a>
<a class="sourceLine" id="cb661-3" data-line-number="3"></a>
<a class="sourceLine" id="cb661-4" data-line-number="4"><span class="kw">confusionMatrix</span>(<span class="kw">predict</span>(train_rf, mnist_<span class="dv">27</span><span class="op">$</span>test),</a>
<a class="sourceLine" id="cb661-5" data-line-number="5">                mnist_<span class="dv">27</span><span class="op">$</span>test<span class="op">$</span>y)<span class="op">$</span>overall[<span class="st">&quot;Accuracy&quot;</span>]</a></code></pre></div>
<pre><code>Accuracy 
    0.79 </code></pre>
<p>Here is what the conditional probabilities look like:</p>
<p><img src="Bayesian-Data-Science_files/figure-html/cond-prob-rf-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>Visualizing the estimate shows that, although we obtain high accuracy, it appears that there is room for improvement by making the estimate smoother. This could be achieved by changing the parameter that controls the minimum number of data points in the nodes of the tree. The larger this minimum, the smoother the final estimate will be. We can train the parameters of the random forest. Below, we use the <strong>caret</strong> package to optimize over the minimum node size. Because, this is not one of the parameters that the <strong>caret</strong> package optimizes by default we will write our own code:</p>
<div class="sourceCode" id="cb663"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb663-1" data-line-number="1">nodesize &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="dv">1</span>, <span class="dv">51</span>, <span class="dv">10</span>)</a>
<a class="sourceLine" id="cb663-2" data-line-number="2">acc &lt;-<span class="st"> </span><span class="kw">sapply</span>(nodesize, <span class="cf">function</span>(ns){</a>
<a class="sourceLine" id="cb663-3" data-line-number="3">  <span class="kw">train</span>(y <span class="op">~</span><span class="st"> </span>., <span class="dt">method =</span> <span class="st">&quot;rf&quot;</span>, <span class="dt">data =</span> mnist_<span class="dv">27</span><span class="op">$</span>train,</a>
<a class="sourceLine" id="cb663-4" data-line-number="4">               <span class="dt">tuneGrid =</span> <span class="kw">data.frame</span>(<span class="dt">mtry =</span> <span class="dv">2</span>),</a>
<a class="sourceLine" id="cb663-5" data-line-number="5">               <span class="dt">nodesize =</span> ns)<span class="op">$</span>results<span class="op">$</span>Accuracy</a>
<a class="sourceLine" id="cb663-6" data-line-number="6">})</a>
<a class="sourceLine" id="cb663-7" data-line-number="7"><span class="kw">qplot</span>(nodesize, acc)</a></code></pre></div>
<p><img src="Bayesian-Data-Science_files/figure-html/acc-versus-nodesize-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>We can now fit the random forest with the optimized minimun node size to the entire training data and evaluate performance on the test data.</p>
<div class="sourceCode" id="cb664"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb664-1" data-line-number="1">train_rf_<span class="dv">2</span> &lt;-<span class="st"> </span><span class="kw">randomForest</span>(y <span class="op">~</span><span class="st"> </span>., <span class="dt">data=</span>mnist_<span class="dv">27</span><span class="op">$</span>train,</a>
<a class="sourceLine" id="cb664-2" data-line-number="2">                           <span class="dt">nodesize =</span> nodesize[<span class="kw">which.max</span>(acc)])</a>
<a class="sourceLine" id="cb664-3" data-line-number="3"></a>
<a class="sourceLine" id="cb664-4" data-line-number="4"><span class="kw">confusionMatrix</span>(<span class="kw">predict</span>(train_rf_<span class="dv">2</span>, mnist_<span class="dv">27</span><span class="op">$</span>test),</a>
<a class="sourceLine" id="cb664-5" data-line-number="5">                mnist_<span class="dv">27</span><span class="op">$</span>test<span class="op">$</span>y)<span class="op">$</span>overall[<span class="st">&quot;Accuracy&quot;</span>]</a></code></pre></div>
<pre><code>Accuracy 
   0.815 </code></pre>
<p>The selected model improves accuracy and provides a smoother estimate.</p>
<p><img src="Bayesian-Data-Science_files/figure-html/cond-prob-final-rf-1.png" width="100%" style="display: block; margin: auto;" /></p>
<p>Note that we can avoid writing our own code by using other random forest implementations as described in the <strong>caret</strong> manual<a href="#fn18" class="footnote-ref" id="fnref18"><sup>18</sup></a>.</p>
<p>Random forest performs better in all the examples we have considered. However, a disadvantage of random forests is that we lose interpretability. An approach that helps with interpretability is to examine <em>variable importance</em>.
To define <em>variable importance</em> we count how often a predictor is used in the individual trees. You can learn more about <em>variable importance</em> in an advanced machine learning book<a href="#fn19" class="footnote-ref" id="fnref19"><sup>19</sup></a>. The <strong>caret</strong> package includes the function <code>varImp</code> that extracts variable importance from any model in which the calculation is implemented. We give an example on how we use variable importance in the next section.</p>
</div>
<div id="exercises-9" class="section level2">
<h2><span class="header-section-number">13.12</span> Exercises</h2>
<p>1. Create a simple dataset where the outcome grows 0.75 units on average for every increase in a predictor:</p>
<div class="sourceCode" id="cb666"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb666-1" data-line-number="1">n &lt;-<span class="st"> </span><span class="dv">1000</span></a>
<a class="sourceLine" id="cb666-2" data-line-number="2">sigma &lt;-<span class="st"> </span><span class="fl">0.25</span></a>
<a class="sourceLine" id="cb666-3" data-line-number="3">x &lt;-<span class="st"> </span><span class="kw">rnorm</span>(n, <span class="dv">0</span>, <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb666-4" data-line-number="4">y &lt;-<span class="st"> </span><span class="fl">0.75</span> <span class="op">*</span><span class="st"> </span>x <span class="op">+</span><span class="st"> </span><span class="kw">rnorm</span>(n, <span class="dv">0</span>, sigma)</a>
<a class="sourceLine" id="cb666-5" data-line-number="5">dat &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">x =</span> x, <span class="dt">y =</span> y)</a></code></pre></div>
<p>Use <code>rpart</code> to fit a regression tree and save the result to <code>fit</code>.</p>
<p>2. Plot the final tree so that you can see where the partitions occurred.</p>
<p>3. Make a scatterplot of <code>y</code> versus <code>x</code> along with the predicted values based on the fit.</p>
<p>4. Now model with a random forest instead of a regression tree using <code>randomForest</code> from the <strong>randomForest</strong> package, and remake the scatterplot with the prediction line.</p>
<p>5. Use the function <code>plot</code> to see if the random forest has converged or if we need more trees.</p>
<p>6. It seems that the default values for the random forest result in an estimate that is too flexible (not smooth). Re-run the random forest but this time with <code>nodesize</code> set at 50 and <code>maxnodes</code> set at 25. Remake the plot.</p>
<p>7. We see that this yields smoother results. Let’s use the <code>train</code> function to help us pick these values. From the <strong>caret</strong> manual<a href="#fn20" class="footnote-ref" id="fnref20"><sup>20</sup></a> we see that we can’t tune the <code>maxnodes</code> parameter or the <code>nodesize</code> argument with <code>randomForest</code>, so we will use the <strong>Rborist</strong> package and tune the <code>minNode</code> argument. Use the <code>train</code> function to try values <code>minNode &lt;- seq(5, 250, 25)</code>. See which value minimizes the estimated RMSE.</p>
<p>8. Make a scatterplot along with the prediction from the best fitted model.</p>
<p>9. Use the <code>rpart</code> function to fit a classification tree to the <code>tissue_gene_expression</code> dataset. Use the <code>train</code> function to estimate the accuracy. Try out <code>cp</code> values of <code>seq(0, 0.05, 0.01)</code>. Plot the accuracy to report the results of the best model.</p>
<p>10. Study the confusion matrix for the best fitting classification tree. What do you observe happening for placenta?</p>
<p>11. Notice that placentas are called endometrium more often than placenta. Note also that the number of placentas is just six, and that, by default, <code>rpart</code> requires 20 observations before splitting a node. Thus it is not possible with these parameters to have a node in which placentas are the majority. Rerun the above analysis but this time permit <code>rpart</code> to split any node by using the argument <code>control = rpart.control(minsplit = 0)</code>. Does the accuracy increase? Look at the confusion matrix again.</p>
<p>12. Plot the tree from the best fitting model obtained in exercise 11.</p>
<p>13. We can see that with just six genes, we are able to predict the tissue type. Now let’s see if we can do even better with a random forest. Use the <code>train</code> function and the <code>rf</code> method to train a random forest. Try out values of <code>mtry</code> ranging from, at least, <code>seq(50, 200, 25)</code>. What <code>mtry</code> value maximizes accuracy? To permit small <code>nodesize</code> to grow as we did with the classification trees, use the following argument: <code>nodesize = 1</code>. This will take several seconds to run. If you want to test it out, try using smaller values with <code>ntree</code>. Set the seed to 1990.</p>
<p>14. Use the function <code>varImp</code> on the output of <code>train</code> and save it to an object called <code>imp</code>.</p>
<p>15. The <code>rpart</code> model we ran above produced a tree that used just six predictors. Extracting the predictor names is not straightforward, but can be done. If the output of the call to train was <code>fit_rpart</code>, we can extract the names like this:</p>
<div class="sourceCode" id="cb667"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb667-1" data-line-number="1">ind &lt;-<span class="st"> </span><span class="op">!</span>(fit_rpart<span class="op">$</span>finalModel<span class="op">$</span>frame<span class="op">$</span>var <span class="op">==</span><span class="st"> &quot;&lt;leaf&gt;&quot;</span>)</a>
<a class="sourceLine" id="cb667-2" data-line-number="2">tree_terms &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb667-3" data-line-number="3"><span class="st">  </span>fit_rpart<span class="op">$</span>finalModel<span class="op">$</span>frame<span class="op">$</span>var[ind] <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb667-4" data-line-number="4"><span class="st">  </span><span class="kw">unique</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb667-5" data-line-number="5"><span class="st">  </span><span class="kw">as.character</span>()</a>
<a class="sourceLine" id="cb667-6" data-line-number="6">tree_terms</a></code></pre></div>
<p>What is the variable importance in the random forest call for these predictors? Where do they rank?</p>
<p>16. Advanced: Extract the top 50 predictors based on importance, take a subset of <code>x</code> with just these predictors and apply the function <code>heatmap</code> to see how these genes behave across the tissues. We will introduce the <code>heatmap</code> function in Chapter <a href="#clustering"><strong>??</strong></a>.</p>

</div>
</div>
<div class="footnotes">
<hr />
<ol start="15">
<li id="fn15"><p><a href="http://www.amazon.com/Mathematical-Statistics-Analysis-Available-Enhanced/dp/0534399428" class="uri">http://www.amazon.com/Mathematical-Statistics-Analysis-Available-Enhanced/dp/0534399428</a><a href="13-examples-of-algorithms.html#fnref15" class="footnote-back">↩</a></p></li>
<li id="fn16"><p><a href="https://web.stanford.edu/~hastie/Papers/ESLII.pdf" class="uri">https://web.stanford.edu/~hastie/Papers/ESLII.pdf</a><a href="13-examples-of-algorithms.html#fnref16" class="footnote-back">↩</a></p></li>
<li id="fn17"><p><a href="https://papers.ssrn.com/sol3/Delivery.cfm/SSRN_ID1759289_code1486039.pdf?abstractid=1759289&amp;mirid=1&amp;type=2" class="uri">https://papers.ssrn.com/sol3/Delivery.cfm/SSRN_ID1759289_code1486039.pdf?abstractid=1759289&amp;mirid=1&amp;type=2</a><a href="13-examples-of-algorithms.html#fnref17" class="footnote-back">↩</a></p></li>
<li id="fn18"><p><a href="http://topepo.github.io/caret/available-models.html" class="uri">http://topepo.github.io/caret/available-models.html</a><a href="13-examples-of-algorithms.html#fnref18" class="footnote-back">↩</a></p></li>
<li id="fn19"><p><a href="https://web.stanford.edu/~hastie/Papers/ESLII.pdf" class="uri">https://web.stanford.edu/~hastie/Papers/ESLII.pdf</a><a href="13-examples-of-algorithms.html#fnref19" class="footnote-back">↩</a></p></li>
<li id="fn20"><p><a href="https://topepo.github.io/caret/available-models.html" class="uri">https://topepo.github.io/caret/available-models.html</a><a href="13-examples-of-algorithms.html#fnref20" class="footnote-back">↩</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="12-cross-validation.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="14-thinking-with-data.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "section",
"scroll_highlight": true
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
